---
title: "MDD interactive heatmaps"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache=FALSE,  message=FALSE, warning = FALSE, fig.height=10, fig.width=10)

suppressMessages(library(tidyverse))
suppressMessages(library(ComplexHeatmap))
library(iheatmapr)
suppressMessages(library(circlize))
library(umap)
```


## Overview{.tabset .tabset-fade}
These figures are derived from the level 4 data downloaded from the LINCS Molecular Deep Dive files on Synapse https://www.synapse.org/#!Synapse:syn2862345/wiki/588244.  

The level 4 data have been filtered for QC issues and normalized to the EGF timepoints.  


```{r standardFigureFunctions}

#' Read an assay's csv or tsv data file 
load_data <- function(assay_name, level = 4){
  file_name <- dir(pattern = paste0("MDD_", assay_name,"_","Level",level,".csv"),recursive = TRUE, full.names = TRUE)
  if(!length(file_name) == 1) stop("There was not one ", assay_name, " data file found")
    
  if(str_detect(file_name, "csv")) {
    df <- read_csv(file_name)
  } else if (str_detect(file_name, "tsv")){
    df <- read_tsv(file_name)
  } else stop("The data file ",file_name, " must be a csv or tsv file type")
} #end of function

get_win_probs <- function(mat, low = .02, hi = .98){
  windsorize_probs <- mat %>%
    quantile(probs = c(low, hi), na.rm = TRUE)
}
  
prep_data <- function(df, md){
    df <- df %>%
    gather(specimenID, value, -feature) %>%
    inner_join(md, by = "specimenID") %>%
    select(feature, specimenName, value) %>%
    spread(specimenName, value) %>%
    mutate(feature = factor(feature, levels = unique(df$feature))) %>%
      arrange(feature,)
  #Set the order of the columns
  condition_order <-  paste(rep(c("ctrl","PBS", "HGF", "OSM", "EGF","BMP2", "IFNG", "TGFB"), each = 30, times = 3), rep(c(0, 1, 4 ,8, 24, 48), each = 5, times = 3), rep(c("C1", "C2", "C3"), each = 240), rep(c("A", "B", "C", "D", "E"), times = 109),  sep = "_") %>%
    intersect(colnames(df))
  
  #Arrange columns by standard condition order
  df <- df %>%
    select(feature, condition_order)
  return(df)
}

prep_annotations <- function(df, md){
      df_ann <- df %>%
      gather("specimenName", "value", -feature) %>%
      inner_join(md, by = "specimenName") %>%
      select(ligand, secondLigand, experimentalTimePoint, replicate, collection) %>%
      distinct() %>%
      rename(Time = experimentalTimePoint,
             Ligand = ligand,
             Ligand2 = secondLigand,
             Replicate = replicate,
             Collection = collection) %>%
      as.data.frame()
      return(df_ann)
}

heatmap_order <- function(df, md, EGF_normed = TRUE){
  
  #convert specimenID column names to specimanName and reorder
  df <- prep_data(df, md)
  if(EGF_normed){
    df <- df %>%
    select(-matches("EGF"))
  }

df_as_matrix <- select(df, -feature) %>%
  as.matrix() %>%
  t %>%
  scale() %>%
  t
  rownames(df_as_matrix) <- df$feature
   #Create the heatmap
  hm <- Heatmap(matrix = df_as_matrix,
          name = "abundance",
          show_column_names = FALSE,
          cluster_columns = FALSE,
          cluster_rows = TRUE,
          na_col = "grey")
  rowOrder <- row_order(hm)
  rowOrder <- rowOrder[length(rowOrder):1]
  return(rowOrder)
}

plot_iheatmap <- function(df, md, assay_name, EGF_normed = TRUE, var_thresh_percentile = 0, win_prob_low = .02, win_prob_up = .98){
  
    df_vars <- df %>%
    select(-feature) %>%
    as.matrix() %>%
    apply(1, var, na.rm = TRUE)
  
  var_thresh <- quantile(df_vars, probs = var_thresh_percentile)
  df <- df[df_vars >= var_thresh,]
  
  #Get a dendrogram from the complex heatmap function
  rowOrder <- heatmap_order(df, md = md, EGF_normed = EGF_normed)

  #convert specimenID column names to specimanName and reorder
  df <- prep_data(df, md)
  
  #Create annotation values
  df_ann <-prep_annotations(df, md) %>%
    select(Collection, Replicate, Time, Ligand2, Ligand)
  
  title_suffix <- ""
  if(EGF_normed){
    title_suffix = ", EGF timecourse normalized"
    df <- df %>%
      select(-matches("EGF"))
    df_ann <- df_ann %>%
      filter(!Ligand == "EGF")
  }
 
  df_as_matrix <- select(df, -feature) %>%
    as.matrix() %>%
    t %>%
  scale() %>%
  t
  rownames(df_as_matrix) <- df$feature
  
windsorize_probs <- get_win_probs(mat = df_as_matrix, win_prob_low, win_prob_up)

  #Create the heatmap
  hm <- main_heatmap(data = df_as_matrix,
                     name = "abundance",
                     zmid = 0,
                     zmin = windsorize_probs[1],  
                     zmax = windsorize_probs[2],
                     row_order = rowOrder) %>%
    #add_row_clustering() %>%
    add_col_annotation(df_ann,
                       colors = list(Ligand = c( "ctrl" = "#7A4A2A",
                                                 "PBS" = "#8dd3c7",
                                                 "HGF" = "#80b1d3",
                                                 "OSM" = "#fdb462",
                                                 "EGF" = "#fb8072",
                                                 "BMP2" = "#b3de69",
                                                 "IFNG" = "#bebada",
                                                 "TGFB" = "#ffd92f"),
                                     Time = c("0" = "white", "1" = "azure1", "4" = "azure2", "8" = "azure3", "24" = "azure4", "48" = "black"), 
                                     Ligand2 = c("EGF" = "#9E79A3", "none" = "#BE7249"),
                                     Replicate = c("A" = '#80b1d3',"B" = '#fdb462',"C" = '#ffd92f',"D" = '#8dd3c7',"E" = '#bebada'),
                                     Collection = c("C1" = "#D79FD4", "C2" = "#3CC3B9", "C3" = "#B9B34C")),
                       size = 0.02)
  hm
}




```

```{r globalVariables}
ligand_cols <- c("ctrl" = "#7A4A2A",
                 "PBS" = "#8dd3c7",
                  "HGF" = "#80b1d3",
                  "OSM" = "#fdb462",
                  "EGF" = "#fb8072",
                  "BMP2" = "#b3de69",
                  "IFNG" = "#bebada",
                  "TGFB" = "#ffd92f")

md <- read_csv("metadata/MDD_sample_annotations.csv")
```

### RPPA{.tabset .tabset-fade}

#### Heatmap

```{r rppa_figures}

RPPA_l1 <- load_data(assay_name = "RPPA", level = 1) %>%
  rename(feature = antibody) %>%
  gather(key = specimenID, value = value, -feature) %>%
  group_by(feature) %>%
  mutate(value = value - median(value, na.rm = TRUE)) %>%
  ungroup() %>%
    spread(key = specimenID, value = value)

```


```{r}
var_thresh_percentile <- 0

plot_iheatmap(RPPA_l1, md, assay_name = "RPPA", EGF_normed = FALSE, var_thresh_percentile = var_thresh_percentile, win_prob_low = .02, win_prob_up = .98)

```

Antibodies with variances above the `r var_thresh_percentile*100` percentile are shown in the heatmaps.  

###  

### IF{.tabset .tabset-fade}

#### Heatmap  

```{r IF_figures, eval = TRUE}

boundedLog2 <- function(x) log2(x + .0001)

immunofluoresence <- load_data(assay_name = "IF", level = 3) %>%
  mutate_if(is.numeric, .funs = boundedLog2) %>%
  filter(!feature == "EdUPositiveProportion")
  
plot_iheatmap(immunofluoresence, md, assay_name = "IF", win_prob_low = .02, win_prob_up = .98)

```

The row features in the heatmap are `r knitr::combine_words(immunofluoresence$feature)`


### cycIF{.tabset .tabset-fade}

#### Heatmap 

```{r cycIF_figures, eval = TRUE}

cycIF <- load_data(assay_name = "cycIF", level = 3) %>%
  filter(str_detect(feature, "_int_")) %>%
  mutate_if(is.numeric, log2)

cycIF_vars <- cycIF %>%
  select(-feature) %>%
  as.matrix() %>%
  apply(1, var)
  
var_thresh <- quantile(cycIF_vars, probs = var_thresh_percentile)
cycIF_hi_vars <- cycIF[cycIF_vars > var_thresh,]

plot_iheatmap(cycIF_hi_vars, md, assay_name = "cycIF")

```


Intensity features included in the heatmap have a variance above the `r var_thresh_percentile` percentile . 


### RNAseq{.tabset .tabset-fade}

#### Heatmap  

```{r rnaSeq_figures, eval = TRUE}
RNAseq <- load_data(assay_name = "RNAseq") %>%
  rename(feature = ensembl_gene_id)

RNAseq_vars <- RNAseq %>%
  select(-feature) %>%
  as.matrix() %>%
  apply(1, var)
  
var_thresh_percentile <- .95
var_thresh <- quantile(RNAseq_vars, probs = var_thresh_percentile)
RNAseq_hi_vars <- RNAseq[RNAseq_vars > var_thresh,]
  
plot_iheatmap(RNAseq_hi_vars, md, assay_name = "RNAseq")

```


Genes included in the heatmap have a variance above the `r var_thresh_percentile` percentile . 



### GCP{.tabset .tabset-fade}

#### Heatmap  

```{r GCP_figures, eval = TRUE}

var_thresh_percentile <- 0


GCP_level2 <- load_data(assay_name = "GCP", level = 2) %>%
  rename(feature = histone) %>%
  gather(key = specimenID, value = value, -feature) %>%
  group_by(feature) %>%
  mutate(value = value - median(value, na.rm = TRUE)) %>%
  ungroup() %>%
    spread(key = specimenID, value = value) %>%
  #mutate(feature = factor(feature, levels = GCP$feature)) %>%
  arrange(feature)
  

plot_iheatmap(GCP_level2, md, assay_name = "GCP", EGF_normed = FALSE, var_thresh_percentile = var_thresh_percentile)

```



### ATACseq{.tabset .tabset-fade}

#### Heatmap  

```{r ATACseq_figures, eval = TRUE}

ATACseq <- load_data(assay_name = "ATACseq") %>%
  rename(feature = peak)

ATACseq_vars <- ATACseq %>%
  select(-feature) %>%
  as.matrix() %>%
  apply(1, var)
  
var_thresh_percentile <- .98
var_thresh <- quantile(ATACseq_vars, probs = var_thresh_percentile)
ATACseq_hi_vars <- ATACseq[ATACseq_vars > var_thresh,]
  
plot_iheatmap(ATACseq_hi_vars, md, assay_name = "ATACseq")

```


Peaks included in the heatmap have a variance above the `r var_thresh_percentile` percentile . 




### L1000{.tabset .tabset-fade}

#### Heatmap  
 
```{r l1000_figures}

L1000 <- load_data("L1000") %>%
  rename(feature = probeset)

```


Genes included in the heatmap have a variance above the `r var_thresh_percentile` percentile . 

```{r}

var_thresh_percentile <- .75
plot_iheatmap(L1000, md, "L1000",var_thresh_percentile = var_thresh_percentile)

```


 
##