---
title: "MDD standard figures"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache=TRUE,  message=FALSE, warning = FALSE, fig.height=10, fig.width=10)

suppressMessages(library(tidyverse))
suppressMessages(library(ComplexHeatmap))
suppressMessages(library(circlize))
library(umap)
library(pheatmap)
library(circlize)
#library(synapser)

```

###Overview
These figures are derived from data downloaded from the LINCS Molecular Deep Dive files on Synapse https://www.synapse.org/#!Synapse:syn2862345/wiki/588244.  

The data have been filtered for QC issues and normalized to the EGF timepoints.  

```{r functions}
#' Read an assay's csv or tsv data file 
load_data <- function(assay_name){
  file_name <- dir("Data",pattern = paste0(assay_name,".*Level4"),recursive = TRUE, full.names = TRUE)
  if(!length(file_name) == 1) stop("There was not one ", assay_name, " data file found")
    
  if(str_detect(file_name, "csv")) {
    df <- read_csv(file_name)
  } else if (str_detect(file_name, "tsv")){
    df <- read_tsv(file_name)
  } else stop("The data file ",file_name, " must be a csv or tsv file type")
} #end of function

####circos functions

parseCorMatrix <- function(x){
  alpha_factor <- 255
  from_condition <- x$Condition
  to_condition <- names(x)[c(-1,-22,-23)]
  #remove metadata from correlations
  cors <- x[c(-1,-22,-23)] %>%
    unlist()
  #remove correlations within the ligand conditions
  cors[str_detect(names(cors), x[["Ligand"]])] <- 0
  #Remove negative correlations
  cors[cors<0] <- 0
  #scale alpha based on correlations
  alpha <- (cors * alpha_factor) %% 256
  alpha <- as.integer(alpha) %>%
    as.character.hexmode()
  col <- paste0(col_ligands[x$Ligand],alpha)
  df <- tibble(From_Condition = x$Condition,
               To_Condition =  names(x)[c(-1,-22,-23)],
               From_sector = str_remove(from_condition, "_.*"),
               To_sector = str_remove(to_condition, "_.*"),
               From_timepoint = str_remove(from_condition, ".*_"),
               To_timepoint = str_remove(to_condition, ".*_"),
               Col = col,
               Cors = as.numeric(cors)) %>%
    mutate(From_timepoint = str_replace(From_timepoint,"^4$",".5"),
           From_timepoint = str_replace(From_timepoint,"^8$","1.5"),
           From_timepoint = str_replace(From_timepoint,"^24$","2.5"),
           From_timepoint = str_replace(From_timepoint,"^48$","3.5"),
           From_timepoint = as.numeric(From_timepoint),
           To_timepoint = str_replace(To_timepoint,"^4$",".5"),
           To_timepoint = str_replace(To_timepoint,"^8$","1.5"),
           To_timepoint = str_replace(To_timepoint,"^24$","2.5"),
           To_timepoint = str_replace(To_timepoint,"^48$","3.5"),
           To_timepoint = as.numeric(To_timepoint),
           From_timepoint_st = From_timepoint - .45,
           From_timepoint_end = From_timepoint + .45,
           To_timepoint_st = To_timepoint - .45,
           To_timepoint_end = To_timepoint + .45)
}

circlize_plot <- function() {
  circos.initializeWithIdeogram(plotType = NULL)
  
  bed = generateRandomBed(nr = 300)
  bed = generateRandomBed(nr = 300, nc = 2)
  circos.genomicTrackPlotRegion(bed,
                                panel.fun = function(region, value, ...) {
                                  circos.genomicPoints(region, value, cex = 0.5, pch = 16, col = 2:3, ...)
                                })
  
  bed = generateRandomBed(nr = 500, nc = 2)
  circos.genomicTrackPlotRegion(bed,
                                panel.fun = function(region, value, ...) {
                                  circos.genomicLines(region, value, col = 4:5, ...)
                                })
  
  bed1 = generateRandomBed(nr = 100)
  bed1 = bed1[sample(nrow(bed1), 20), ]
  bed2 = generateRandomBed(nr = 100)
  bed2 = bed2[sample(nrow(bed2), 20), ]
  
  circos.genomicLink(bed1, bed2, col = col_fun(bed1[[4]]))
  
  circos.clear()
}
addLink <- function(x){
  cor_thresh <- .45
  if(x["Cors"] > cor_thresh) {
    lwd <- max(0, as.numeric(x["Cors"])-cor_thresh)*30
    #cat("lwd ",lwd, "col ", x["Col"])
    circos.link(x["From_sector"], as.numeric(x["From_timepoint"]), x["To_sector"], as.numeric(x["To_timepoint"]), h = 0.4, lwd = lwd, col =x["Col"])
  }
}

prepare_data <- function(x){
  df <- x %>%
    data.frame() %>%
    rownames_to_column(var ="Condition") %>%
    mutate(Time = str_remove(Condition, ".*_"),
           Time = str_replace_all(Time, "^4$","1"),
           Time = str_replace_all(Time, "^8$","2"),
           Time = str_replace_all(Time, "^24$","3"),
           Time = str_replace_all(Time, "^48$","4"),
           Ligand = str_remove(Condition, "_.*")
           )
}

draw_one_track <- function(title_text){
  circos.par("track.height" = 0.2,
             "points.overflow.warning" = FALSE)
  circos.initialize(factors = as.factor(rppa$Ligand), xlim =c(0,4), sector.width = 20)

  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    xlim = CELL_META$xlim
    ylim = CELL_META$ylim
    breaks = seq(xlim[1], xlim[2], by = 1)
    n_breaks = length(breaks)
    circos.rect(breaks[-n_breaks], rep(ylim[1], n_breaks - 1),
                breaks[-1], rep(ylim[2], n_breaks - 1),
                col = track_col_names,
                border = NA)
    circos.text(CELL_META$xcenter,
                CELL_META$cell.ylim[2] + uy(3, "mm"),
                CELL_META$sector.index,
                col = c("black","red","black", "red", "red"))
  }) 
  
  title(title_text,
        outer = TRUE,
        cex.main = 1,
        line = -1)
  
  legend(-1.1,
         -.9,
         "TGFB, BMP2 and IFNG\n are paired with EGF",
         bty = "n",
         cex = .6)
}

get_win_probs <- function(df, low = .02, hi = .98){
  windsorize_probs <- df %>%
    select(-feature) %>%
    unlist %>%
    quantile(probs = c(low, hi))
}
  
prep_data <- function(df){
    df <- df %>%
    gather(specimenID, value, -feature) %>%
    inner_join(md, by = "specimenID") %>%
    select(feature, specimenName, value) %>%
    spread(specimenName, value)
  #Set the order of the columns
  condition_order <-  paste(rep(c("PBS","BMP2", "IFNG", "TGFB", "HGF", "OSM", "EGF"), each = 25, times = 3), rep(c(1, 4 ,8, 24, 48), each = 5, times = 3), rep(c("C1", "C2", "C3"), each = 175), rep(c("A", "B", "C", "D", "E"), times = 105),  sep = "_") %>%
    intersect(colnames(df))
  
  #Arrange columns by standard condition order
  df <- df %>%
    select(feature, condition_order)
  return(df)
}

prep_annotaitions <- function(df, md){
      df_ann <- df %>%
      gather("specimenName", "value", -feature) %>%
      inner_join(md, by = "specimenName") %>%
      select(ligand, secondLigand, experimentalTimePoint, replicate) %>%
      distinct() %>%
      rename(Time = experimentalTimePoint,
             Ligand = ligand,
             Ligand2 = secondLigand) %>%
      as.data.frame()
      return(df_ann)
}
 #Create the heatmap column annotations
  create_top_annotations <- function(df){
    df_ann <- prep_annotaitions(df, md)
    
    haTop <- HeatmapAnnotation(df = df_ann,
                               col = list(Ligand = c("PBS" = "#8dd3c7", "BMP2" = "#ffffb3", "IFNG" = "#bebada", "TGFB" = "#b3de69", "HGF" = "#80b1d3", "OSM" = "#fdb462", "EGF" = "#fb8072"),
                                          Time = c("1" = "azure1", "4" = "azure2", "8" = "azure3", "24" = "azure4", "48" = "black"), 
                                          Ligand2 = c("EGF" = "#fc8d62", "none" = "#8da0cb")))
    return(haTop)
  }
  
#####
plot_heatmap <- function(df, md, assay_name, show_row_names = FALSE){
  
  windsorize_probs <- get_win_probs(df, .01, .99)
  
  #convert specimenID column names to specimanName and reorder
  df <- prep_data(df)
  
  #Create the heatmap
  Heatmap(matrix = select(df, -feature) %>%
            as.matrix(),
          name = "abundance",
          top_annotation = create_top_annotations(df),
          show_row_names = show_row_names,
          show_column_names = FALSE,
          cluster_columns = FALSE,
          col = colorRamp2(c(windsorize_probs[1], 0, windsorize_probs[2]), c("blue", "white", "red")),
          na_col = "grey")
}

plot_umap <- function(df, md, assay_name){
  #Generate a 2d scatterplot of the UMAP dimensions
  #Use size to represent timepoint, dot in center to represent ligand_2 == EGF

  ligand_cols <- c("PBS" = "#8dd3c7",
                   "BMP2" = "#ffffb3",
                   "IFNG" = "#bebada",
                   "TGFB" = "#b3de69",
                   "HGF" = "#80b1d3",
                   "OSM" = "#fdb462",
                   "EGF" = "#fb8072")
  
   ligand_2_cols <- c("EGF" = "#ff0000",
                     "None" = "#00000000")
  windsorize_probs <- get_win_probs(df, .01, .99)
  
  #convert specimenID column names to specimanName and reorder
  df <- prep_data(df)
  
   #Create annotation values
  df_ann <-prep_annotaitions(df, md)

   df_mat <- df %>%
    select(-feature) %>%
     t() %>%
  as.matrix()
  colnames(df_mat) <- df$feature

  df_UMAP <- umap(df_mat)$layout %>%
  data.frame() %>%
  rename(UMAP_1 = X1,
         UMAP_2 = X2) %>%
    cbind(df_ann)
  
p <- ggplot(df_UMAP, aes(x = UMAP_1,
                         y = UMAP_2,
                         size = Time,
                         colour = Ligand,
                         fill = Ligand2)) +
    geom_point() +
  geom_point(shape = 21, stroke = 4) +
  scale_color_manual( values = ligand_cols) +
  scale_fill_manual( values = ligand_2_cols) +
  scale_size(breaks = c(1,4,8,24,48), trans = "exp") +
  labs(title = paste("UMAP embedding of ", assay_name, " data"))
p
  
}

plot_pca <- function(df){
    #Generate a 2d scatterplot of the the first two principal components
  #Use size to represent timepoint, a dot in center to represent ligand_2 == EGF
 #temporary function for Ligand_2 metadata
  label_ligand_2 <- function(x) {
    ligand_2 <- rep("None", times = length(x))
    ligand_2[x %in% c("BMP2", "IFNG", "TGFB")] <- "EGF"
    return(ligand_2)
  }
  
  ligand_cols <- c("PBS" = "#8dd3c7",
                   "BMP2" = "#ffffb3",
                   "IFNG" = "#bebada",
                   "TGFB" = "#b3de69",
                   "HGF" = "#80b1d3",
                   "OSM" = "#fdb462",
                   "EGF" = "#fb8072")
  
   ligand_2_cols <- c("EGF" = "#ff0000",
                     "None" = "#00000000")

   #Create annotation values
  df_ann <- df %>%
    select(-Feature) %>%
    colnames() %>%
    enframe(value = "Condition") %>%
    mutate(Ligand = str_remove(Condition, "_.*"),
           Timepoint = str_remove(Condition, ".*_"),
           Timepoint = as.numeric(Timepoint),
           Ligand_2 = label_ligand_2(Ligand)) %>%
    select(-Condition, -name) %>%
    as.data.frame()

   df_mat <- df %>%
    select(-Feature) %>%
    t() %>%
  as.matrix()
  colnames(df_mat) <- df$Feature

  df_pca <- prcomp(df_mat)$x %>%
  data.frame() %>%
    cbind(df_ann)
  
p <- ggplot(df_pca, aes(x = PC1,
                         y = PC2,
                         size = Timepoint,
                         colour = Ligand,
                         fill = Ligand_2)) +
    geom_point() +
  geom_point(shape = 21, stroke = 4) +
  scale_color_manual( values = ligand_cols) +
  scale_fill_manual( values = ligand_2_cols) +
  scale_size(breaks = c(1,4,8,24,48), trans = "exp") +
  labs(title = paste("PCA of ", assay_name, " data"))
p
  
}

plot_introspect <- function(df, assay_name){
  #calculate correlations across the condtions and show in a heatmap
#temporary function for Ligand_2 metadata
  label_ligand_2 <- function(x) {
    ligand_2 <- rep("None", times = length(x))
    ligand_2[x %in% c("BMP2", "IFNG", "TGFB")] <- "EGF"
    return(ligand_2)
  }
  
  windsorize_probs <- df %>%
    select(-Feature) %>%
  unlist %>%
  quantile(probs = c(.005,.995))
  
  #Set the order of the columns
  condition_order <-  paste(rep(c("PBS","BMP2", "IFNG", "TGFB", "HGF", "OSM", "EGF"), each = 5), rep(c(1, 4 ,8, 24, 48), times = 7), sep = "_") %>%
    intersect(colnames(df))
  
  #Arrange columns by standard condition order
  df <- df %>%
    select(Feature, condition_order)
  #Create the heatmap column annotations
  df_ann <- colnames(df) %>%
    intersect(condition_order) %>%
    enframe(value = "Condition") %>%
    mutate(Ligand = str_remove(Condition, "_.*"),
           Timepoint = str_remove(Condition, ".*_"),
           Timepoint = as.numeric(Timepoint),
           Ligand_2 = label_ligand_2(Ligand)) %>%
    select(-Condition, -name) %>%
    as.data.frame()
  
  haTop = HeatmapAnnotation(df = df_ann,
                            col = list(Ligand = c("PBS" = "#8dd3c7", "BMP2" = "#ffffb3", "IFNG" = "#bebada", "TGFB" = "#b3de69", "HGF" = "#80b1d3", "OSM" = "#fdb462", "EGF" = "#fb8072"),
                                       Timepoint = c("1" = "azure1", "4" = "azure2", "8" = "azure3", "24" = "azure4", "48" = "black"), 
                                       Ligand_2 = c("EGF" = "#fc8d62", "None" = "#8da0cb")))
  
  haRow <- rowAnnotation(df = df_ann,
                            col = list(Ligand = c("PBS" = "#8dd3c7", "BMP2" = "#ffffb3", "IFNG" = "#bebada", "TGFB" = "#b3de69", "HGF" = "#80b1d3", "OSM" = "#fdb462", "EGF" = "#fb8072"),
                                       Timepoint = c("1" = "azure1", "4" = "azure2", "8" = "azure3", "24" = "azure4", "48" = "black"), 
                                       Ligand_2 = c("EGF" = "#fc8d62", "None" = "#8da0cb")))
  
  #Create the heatmap
  Heatmap(matrix = cor(select(df, -matches("Feature"))),
          name = "abundance",
          top_annotation = haTop,
          right_annotation = haRow,
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = FALSE,
          cluster_columns = FALSE,
          col = colorRamp2(c(windsorize_probs[1], 0, windsorize_probs[2]), c("blue", "white", "red")),
          na_col = "grey")
}

plot_circos <- function(df){
 ####DEBUG refactor to match new data sturcture###### 
#Assign colors to the ligands 
col_ligands <- c("PBS" = "#8dd3c7", "BMP2" = "#ffffb3", "IFNG" = "#bebada", "TGFB" = "#b3de69", "HGF" = "#80b1d3", "OSM" = "#fdb462", "EGF" = "#fb8072", "ctrl" = "#ffffff66")
names(col_ligands) <- unique(rppa$Ligand)

#Assign colors to conditions in the track
track_col_names <- c("1" = "azure1", "4" = "azure2", "8" = "azure3", "24" = "azure4", "48" = "black") 

res <- draw_one_track(title_text = "RPPA Correlations")
df <- rppa
for(i in 1:nrow(df)){
  foo <- parseCorMatrix(df[i,])
  res <- apply(foo, 1, addLink)
}
circos.clear()

}
```


```{r rppa_figures}

md <- read_tsv("metadata/MCF10A_MDD_sample_annotations.tsv")

df <- load_data("RPPA") %>%
  rename(feature = antibody)

plot_heatmap(df, md, assay_name = "RPPA")

plot_umap(df, md, assay_name = "RPPA")

#plot_pca(df, md, assay_name = "RPPA")

#plot_introspect(df, md, assay_name = "RPPA")

```

```{r IF_figures, eval = FALSE}

immunofluoresence <- load_data("IF")

#Select data and translate from IF terms to common terms
df <- immunofluoresence %>%
  mutate(Collection_condition = paste(Collection, Ligand, TimePoint, sep = "_")) %>%
  select(Collection_condition, Replicate, matches("EGFNorm")) %>%
  group_by(Collection_condition) %>%
  summarise_if(is.numeric, mean) %>%
  ungroup() %>%
  gather(key = "Feature") %>%

  gather(-Collection, -Ligand, -TimePoint)
  rename(Feature = Antibody,
         Condition = Ligand_TimePoint,
         Value = EGFNorm) %>%
  spread(key = Feature, value = Value) %>%
  filter(!Condition == "ctrl_0")

res <- plot_heatmap(df, assay = "RPPA")
res

```

```{r cycIF_figures, eval = FALSE}


```

```{r rnaSeq_figures, eval = FALSE}


```

```{r GCP_figures, eval = FALSE}


```

```{r ATACseq_figures, eval = FALSE}


```

```{r l1000_figures, eval = FALSE}


```