---
title: "Chromatin Accessibility in Expressed/Unexpressed Genes"
author: "Daniel Derrick"
date: "5/12/2020"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    df_print: paged
  html_notebook:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r librariesPaths, include = FALSE}
library(genomation)
library(biomaRt)
library(tidyverse)
library(cmapR)
library(ComplexHeatmap)
library(cowplot)
library(ggpubr)
library(circlize)
library(goseq)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(GenomicFeatures)
library(DT)
library(xlsx)

theme_set(theme_cowplot())

RNAscript <- "R/MDD_import_RNAseq_ensembl.R"
RNAseqResultsFile <- "RNAseq/MDD_RNAseq_shrunkenResults.Rdata"
transcriptsFile <- "misc/MDD_RNAseq_mostAbundantTranscripts.txt"
ATACseqSAFile   <- "ATACseq/Metadata/MDD_ATACseq_sampleMetadata.csv"
ATACseqAccTable <- "ATACseq/Data/misc/promoter_accessibility_table.csv"
HLA_filtFile <- "RNAseq/misc/HLA_geneList.csv"
promoterDir <- "ATACseq/Data/misc"
colScript <- "R/MDD_importColors_pretty.R"
bwDir <- "ATACseq/Data/Level0/bigWig"
outDirPlots <- "plots/RNAseqATACseq"
heatmapFunctionScript <- "R/MDD_RNAseq_heatmapFunctions.R"

RNAseq_lfc_threshold  <- 1.5
RNAseq_padj_threshold <- 0.01

# Functions for ATAC-seq data
getProm <- function(txdb, annoTable, buffer = 1500, genesOnly = FALSE) {
  txdb <- keepStandardChromosomes(txdb)
  txdb <- dropSeqlevels(txdb, c("chrY", "chrM"))
  
  if (genesOnly) {
    txdb <- genes(txdb)
    prom <- promoters(txdb, upstream = buffer, downstream = buffer)
    prom$ensembl_gene_id <- annoTable[match(prom$gene_id, 
                                            annoTable$entrezgene), 1]
  } else {
    prom <- promoters(txdb, upstream = buffer, downstream = buffer)
    prom$tx_name         <- str_remove(prom$tx_name, "[.][0-9]+")
    prom$ensembl_gene_id <- annoTable[match(prom$tx_name, 
                                            annoTable$ensembl_transcript_id), ]$ensembl_gene_id
  }
  
  filt <- is.na(prom$ensembl_gene_id)
  prom <- prom[!filt]
  
  return(prom)
}

combineMat <- function(geneMat, combine) {
  dumSum <- function(x, y) {
    z <- x + y
    return(z)
  }
  
  geneMatAvg <-
    lapply(combine, function(x) {
      # print(x)
      x_sum <-
        Reduce(dumSum, geneMat[x])
      x_avg <- (x_sum/length(x))
      return(x_avg)
    })
  
  names(geneMatAvg) <- names(combine)
  geneMatAvg <- as(geneMatAvg, "ScoreMatrixList")
  return(geneMatAvg)
}


HeatmapL3C <- function(mat, ca = ha.RNA.L3, featureName = "Z-score", showRow = FALSE, clust = FALSE, ...) {
  Heatmap(mat, 
          top_annotation = ca,
          cluster_columns = clust,
          show_row_names = showRow,
          cluster_row_slices = FALSE,
          row_names_gp = gpar(fontsize = 7),
          show_column_names = FALSE,
          name = featureName,
          col = colorRamp2(c(-3, 0, 3), c("blue", "white", "red")),
          ...)
}


getProm <- function(txdb, annoTable, buffer = 1500, genesOnly = FALSE) {
  txdb <- keepStandardChromosomes(txdb)
  txdb <- dropSeqlevels(txdb, c("chrY", "chrM"))
  
  if (genesOnly) {
    txdb <- genes(txdb)
    prom <- promoters(txdb, upstream = buffer, downstream = buffer)
    prom$ensembl_gene_id <- aT[match(prom$gene_id, aT$entrezgene), 1]
  } else {
    prom <- promoters(txdb, upstream = buffer, downstream = buffer)
    prom$tx_name         <- str_remove(prom$tx_name, "[.][0-9]+")
    prom$ensembl_gene_id <- aT[match(prom$tx_name, aT$ensembl_transcript_id), 1]
  }
  
  filt <- is.na(prom$ensembl_gene_id)
  prom <- prom[!filt]
  
  return(prom)
}

geneGrangestoTSS <- function(x, buffer = 1000) {
  strands <- as.vector(strand(x))
  
  end(x[strands == "+"])   <- start(x[strands == "+"]) + buffer
  start(x[strands == "+"]) <- start(x[strands == "+"]) - buffer
  
  start(x[strands == "-"]) <- end(x[strands == "-"]) - buffer
  end(x[strands == "-"])   <- end(x[strands == "-"]) + buffer
  
  x
}

combineMat <- function(geneMat, combine) {
  dumSum <- function(x, y) {
    z <- x + y
    return(z)
  }
  
  geneMatAvg <-
    lapply(combine, function(x) {
      print(x)
      x_sum <-
        Reduce(dumSum, geneMat[x])
      x_avg <- (x_sum/length(x))
      return(x_avg)
    })
  
  names(geneMatAvg) <- names(combine)
  geneMatAvg <- as(geneMatAvg, "ScoreMatrixList")
  return(geneMatAvg)
}


source(colScript)
###############################################################################
```

```{r}
source(heatmapFunctionScript)
```

# Introduction

In this document, expressed/unexpressed gene lists are calculated for each
`experimentalCondition`. 

The expressed/unexpressed gene lists are then 
joined to RNAseq DE gene results.

This will then be joined to promoter accessibility in the ctrl condition.

## Identifying expressed/unexpressed genes

Returning to the minimum expression filter of log2fpkm ≥ 0.05 in ≥2 replicates 
of a condition.

```{r}
source(RNAscript)

exprFilt <-
  RNAseqL3 %>% 
  data.frame %>% 
  rownames_to_column("ensembl_gene_id") %>% 
  pivot_longer(-ensembl_gene_id, names_to = "specimenID", values_to = "log2fpkm") %>% 
  left_join(sa.RNA.L3, by = "specimenID") %>% 
  group_by(experimentalCondition, ensembl_gene_id) %>% 
  summarize(expressed = sum(log2fpkm >= 0.5) >= 2) %>% 
  ungroup

exprFilt %>% 
  filter(expressed) %>% 
  group_by(experimentalCondition, expressed) %>% 
  summarize(expressedGenes = n()) %>% 
  ungroup %>% 
  dplyr::select(-expressed)

exprFilt %>% 
  filter(expressed) %>% 
  group_by(experimentalCondition, expressed) %>% 
  summarize(expressedGenes = n()) %>% 
  ungroup %>% 
  left_join(distinct(sa.RNA.L3, experimentalCondition, .keep_all = TRUE)) %>% 
  mutate(experimentalCondition = fct_relevel(experimentalCondition, 
                                             unique(sa.RNA.L3$experimentalCondition))
         ) %>% 
  ggplot(aes(experimentalCondition, expressedGenes)) +
  geom_col(aes(fill = Ligand)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = col$Ligand) +
  ylab("Expressed Genes")
```

## Filtering expressed genes to protein-coding genes

Filtering RNAseq data to only protein-coding genes.

```{r, fig.height = 10, fig.width = 10}
exprFilt %>% 
  mutate(experimentalCondition = fct_relevel(experimentalCondition, 
                                             unique(sa.RNA.L3$experimentalCondition))) %>% 
  left_join(at.RNA) %>% 
  mutate(gene_biotype = as.factor(gene_biotype)) %>% 
  mutate(geneType = fct_collapse(gene_biotype,
                                 pseudogene = grep("pseudogene", levels(gene_biotype), 
                                                   value = TRUE),
                                 other = grep("_gene", levels(gene_biotype), 
                                              value = TRUE))) %>%
  filter(expressed) %>% 
  group_by(geneType, experimentalCondition) %>% 
  summarize(n = n()) %>%
  ungroup %>% 
  arrange(desc(n)) %>% 
  mutate(geneType = fct_drop(geneType)) %>% 
  mutate(geneType = fct_inorder(geneType)) %>% 
  left_join(sa.RNA.L3 %>% 
              mutate(experimentalCondition = fct_inorder(as.factor(experimentalCondition))) %>% 
              distinct(experimentalCondition, .keep_all = TRUE)) %>% 
  ggplot(aes(experimentalCondition, n)) +
  geom_col(aes(fill = Ligand)) +
  facet_wrap(~geneType, ncol = 3, scale = "free_y") +
  theme(axis.text.x = element_text(size = 8, angle = 90, hjust = 1)) +
  scale_fill_manual(values = col$Ligand) +
  ggtitle("# of Expressed Genes Per Category")

exprFilt <- 
  exprFilt %>% 
  mutate(experimentalCondition = fct_relevel(experimentalCondition, 
                                             unique(sa.RNA.L3$experimentalCondition))) %>% 
  left_join(at.RNA) %>% 
  mutate(gene_biotype = as.factor(gene_biotype)) %>% 
  mutate(geneType = fct_collapse(gene_biotype,
                                 pseudogene = grep("pseudogene", levels(gene_biotype), 
                                                   value = TRUE),
                                 other = grep("_gene", levels(gene_biotype), 
                                              value = TRUE))) %>%
  filter(geneType == "protein_coding")
```

```{r}
exprFilt %>% 
  filter(expressed) %>% 
  group_by(experimentalCondition, expressed) %>% 
  summarize(expressedGenes = n()) %>% 
  ungroup %>% 
  left_join(distinct(sa.RNA.L3, experimentalCondition, .keep_all = TRUE)) %>% 
  mutate(experimentalCondition = fct_relevel(experimentalCondition, 
                                             unique(sa.RNA.L3$experimentalCondition))
         ) %>% 
  ggplot(aes(experimentalCondition, expressedGenes)) +
  geom_col(aes(fill = Ligand)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = col$Ligand) +
  ylab("Expressed\nProtein-Coding Genes")
```


### Importing differentially expressed genes.

Loading DESeq2 results table, joining with expression filter. Annotations as follows:

|Annotation|log2FoldChange Threshold|FDR-adjusted p-value threshold|Expression in condition | Expression in CTRL|
|---|---|---|---|---|
|DE     | ≥ 1.5 or ≤ -1.5| ≤ 0.01 | NA | NA |
|Induced| ≥ 1.5          | ≤ 0.01  | TRUE | FALSE |
|Repressed| ≤ -1.5       | ≤ 0.01  | FALSE | TRUE |

```{r}
load(RNAseqResultsFile)

RNAseq_resExpr <-
  Reduce(bind_rows, res_RNA_df) %>% 
  mutate(padj = replace_na(padj, 1)) %>% 
  dplyr::select(-hgnc_symbol) %>% 
  right_join(filter(exprFilt, experimentalCondition != "ctrl_0"),
                    by = c("experimentalCondition", "ensembl_gene_id")) %>% 
  left_join(exprFilt %>% 
              filter(experimentalCondition == "ctrl_0") %>% 
              dplyr::rename(expressedCtrl = expressed) %>% 
              dplyr::select(ensembl_gene_id, expressedCtrl),
            by = "ensembl_gene_id") %>% 
  mutate(DE = (expressed &
                 abs(log2FoldChange) >= RNAseq_lfc_threshold &
                 padj <= RNAseq_padj_threshold),
         Induced = (expressed & 
                      !expressedCtrl &
                      log2FoldChange >= RNAseq_lfc_threshold &
                      padj <= RNAseq_padj_threshold),
         Repressed = (!expressed &
                        expressedCtrl &
                        log2FoldChange <= -RNAseq_lfc_threshold &
                        abs(log2FoldChange) >= RNAseq_padj_threshold)
  ) %>% 
  mutate(DE = DE | Repressed)

RNAseq_resExpr %>% 
  filter(DE) %>% 
  left_join(distinct(sa.RNA.L3, experimentalCondition, .keep_all = TRUE)) %>% 
  mutate(experimentalCondition = fct_relevel(experimentalCondition, 
                                             unique(sa.RNA.L3$experimentalCondition))
         ) %>% 
  group_by(experimentalCondition) %>% 
  summarize("n_DE_genes" = n()) 

RNAseq_resExpr %>% 
  filter(DE) %>% 
  left_join(distinct(sa.RNA.L3, experimentalCondition, .keep_all = TRUE)) %>% 
  mutate(experimentalCondition = fct_relevel(experimentalCondition, 
                                             unique(sa.RNA.L3$experimentalCondition))
         ) %>% 
  ggplot(aes(experimentalCondition)) +
  geom_bar(aes(fill = Ligand)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_manual(values = col$Ligand) +
  ggtitle("# of DE Protein-Coding Genes per Condition") 
```

### Removing HLA region genes

Genes in HLA region will lack ATAC-seq accessibility due to alignment to alternative
chromosomes; flagging and removing these genes from further analysis. **98 unique
genes are removed.**

```{r}
HLAfilt <- read.csv(HLA_filtFile)

RNAseq_resExpr <- 
  RNAseq_resExpr %>% 
  left_join(HLAfilt) %>% 
  mutate(HLA = replace_na(HLA, FALSE))

RNAseq_resExpr %>% 
  filter(DE) %>% 
  filter(HLA) %>% 
  left_join(distinct(sa.RNA.L3, experimentalCondition, .keep_all = TRUE)) %>% 
  mutate(experimentalCondition = fct_relevel(experimentalCondition, 
                                             unique(sa.RNA.L3$experimentalCondition))
         ) %>% 
  group_by(experimentalCondition, HLA) %>% 
  summarize(n = n())

RNAseq_resExpr %>% 
  filter(DE) %>% 
  left_join(distinct(sa.RNA.L3, experimentalCondition, .keep_all = TRUE)) %>% 
  mutate(experimentalCondition = fct_relevel(experimentalCondition, 
                                             unique(sa.RNA.L3$experimentalCondition))
         ) %>% 
  group_by(experimentalCondition, HLA) %>% 
  ggplot(aes(experimentalCondition)) +
  geom_bar(aes(fill = HLA)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_brewer(type = "qual", palette = 2) +
  ggtitle("# of DE Genes per Condition", subtitle = "HLA-region genes marked")

RNAseq_resExpr <- 
  RNAseq_resExpr %>% 
  filter(!HLA)
```

Most HLA region DE genes are in IFNG conditions.

### Joining to table of accessible promoters

```{r}
accTable <- 
  read.csv(ATACseqAccTable, stringsAsFactors = FALSE) %>% 
  dplyr::select(-expressed, -gene_biotype, -hgnc_symbol)

RNAseq_resExprAcc <- 
  RNAseq_resExpr %>% 
  left_join(accTable) %>% 
  mutate(accessible = replace_na(accessible, FALSE))
```

## DE gene summary - all conditions

### Total # of DE Genes

"Across the ligand treatments we found that XX genes had XX log fold change in at 
least one of the treatment conditions."

```{r}
RNAseq_resExprAcc %>% 
  group_by(ensembl_gene_id) %>% 
  summarize(anyDE = any(DE, na.rm = TRUE)) %>% 
  ungroup %>% 
  group_by(anyDE) %>% 
  summarize(n = n())
```

**Across the ligand treatments we found that 3158 unique protein-coding
genes had at least ±1.5 log fold change in at least one of the treatment conditions.**

### Total # of induced (from 0) / repressed (to 0) genes

"XX of these genes had expression reduced to zero, and XX genes were newly expressed.""

```{r}
RNAseq_resExprAcc %>% 
  group_by(ensembl_gene_id) %>% 
  summarize(anyInduced = any(Induced, na.rm = TRUE),
            anyRepressed = any(Repressed, na.rm = TRUE)) %>% 
  ungroup() %>% 
  group_by(anyInduced, anyRepressed) %>% 
  summarize(n = n())
```

**674/3158 (21.3%) of these genes had expression reduced to zero, 
and 603/3158 (19.1%) genes were newly expressed.**

### Accessibility of induced genes

"Analysis of chromatin TSS showed that XX% of the newly expressed genes had closed chromatin in the CTRL condition"

```{r}
RNAseq_resExprAcc %>% 
  filter(Induced | Repressed) %>% 
  group_by(ensembl_gene_id) %>% 
  summarize(anyInduced = any(Induced),
            anyRepressed = any(Repressed),
            accessible = any(accessible)) %>% 
  ungroup %>% 
  group_by(anyInduced, anyRepressed, accessible) %>% 
  summarize(n = n())
```

**Analysis of chromatin TSS showed that 67.7% (408/603) of the newly expressed
protein-coding genes had closed chromatin in the CTRL condition**

## DE gene summary - IFNG_48

### Total # of DE Genes

"IFNG_48 had XX unique protein-coding
genes with at least ±1.5 log fold change."

```{r}
RNAseq_resExprAcc %>% 
  filter(experimentalCondition == "IFNG_48") %>%
  group_by(ensembl_gene_id) %>% 
  summarize(anyDE = any(DE, na.rm = TRUE)) %>% 
  ungroup %>% 
  group_by(anyDE) %>% 
  summarize(n = n())
```

"IFNG_48 had 1119 unique protein-coding
genes with at least ±1.5 log fold change."

### Total # of induced (from 0) / repressed (to 0) genes

"XX of these genes had expression reduced to zero, and XX genes were newly expressed.""

```{r}
RNAseq_resExprAcc %>% 
  filter(experimentalCondition == "IFNG_48") %>%
  group_by(ensembl_gene_id) %>% 
  summarize(anyInduced = any(Induced, na.rm = TRUE),
            anyRepressed = any(Repressed, na.rm = TRUE)) %>% 
  ungroup() %>% 
  group_by(anyInduced, anyRepressed) %>% 
  summarize(n = n())
```

**179/1119 (16.0%) of these genes had expression reduced to zero, 
and 233/1119 (20.8%) genes were newly expressed.**

### Accessibility of induced genes

"Analysis of chromatin TSS showed that XX% of the newly expressed genes had closed chromatin in the CTRL condition"

```{r}
RNAseq_resExprAcc %>% 
  filter(experimentalCondition == "IFNG_48") %>%
  filter(Induced | Repressed) %>% 
  group_by(ensembl_gene_id) %>% 
  summarize(anyInduced = any(Induced),
            anyRepressed = any(Repressed),
            accessible = any(accessible)) %>% 
  ungroup %>% 
  group_by(anyInduced, anyRepressed, accessible) %>% 
  summarize(n = n())
```

**Analysis of chromatin TSS showed that 67.4% (157/233) of the newly expressed
protein-coding genes had closed chromatin in the CTRL condition**

## Plots

### Donut plot: Protein-coding genes

```{r}
forDonut <-
  RNAseq_resExprAcc %>% 
  filter(geneType == "protein_coding") %>% 
  group_by(ensembl_gene_id) %>% 
  summarize(Category = case_when(any(Induced, na.rm = TRUE) ~ "Induced",
                                 any(Repressed, na.rm = TRUE) ~ "Repressed",
                                 any(DE, na.rm = TRUE) ~ "Differentially Expressed",
                                 !any(DE, na.rm = TRUE) & any(expressed, na.rm = TRUE) ~ "Expressed, no change",
                                 !any(expressed) ~ "Not Expressed")) %>% 
  ungroup %>% 
  group_by(Category) %>% 
  summarize(n = n()) %>% 
  ungroup %>% 
  mutate(fraction = n/sum(n))
forDonut$ymax = cumsum(forDonut$fraction)
forDonut$ymin = c(0, head(forDonut$ymax, n = -1))

forDonut <-
  forDonut %>% 
  mutate(Category = fct_relevel(as.factor(Category), c("Not Expressed", "Expressed, no change", "Differentially Expressed", 
                                                       "Induced", "Repressed")))

a <- ggdonutchart(forDonut, "n", "n", 
             fill = "Category") +
  ggtitle("19538 Protein-Coding Genes") +
  scale_fill_manual(values = c("Not Expressed" = "gray30",
                               "Expressed, no change" = "gray95",
                               "Differentially Expressed" = "darkorchid2",
                               "Induced" = "firebrick3",
                               "Repressed" = "steelblue3"))
a <- ggpar(a, legend = "right")
a

# if(!dir.exists(outDirPlots)) {dir.create(outDirPlots)}
# 
# pdf(sprintf("%s/MDD_RNAseq_geneCategoryDonutPlot.pdf", outDirPlots), height = 5, width = 6)
# a
# dev.off()
```

### Heatmap: Induced & Repressed Genes

```{r}
inducedGenes <- RNAseq_resExprAcc %>% 
  filter(Induced) %>% 
  distinct(ensembl_gene_id) %>% 
  pull(ensembl_gene_id)

repressedGenes <- RNAseq_resExprAcc %>% 
  filter(Repressed) %>% 
  distinct(ensembl_gene_id) %>% 
  pull(ensembl_gene_id)

Heatmap(RNAseqL3[c(inducedGenes, repressedGenes), ], 
        cluster_columns = FALSE, split = rep(c("Induced", "Repressed"), 
                                             c(length(inducedGenes), length(repressedGenes))),
        show_row_names = FALSE,
        top_annotation = ha.RNA.L3,
        show_column_names = FALSE,
        col = colorRamp2(c(0, .5, 4), c("blue", "white", "red")),
        name = "log2(fpkm + 1)")

Heatmap(RNAseqL3Z[c(inducedGenes, repressedGenes), ], 
        cluster_columns = FALSE, split = rep(c("Induced", "Repressed"), 
                                             c(length(inducedGenes), length(repressedGenes))),
        show_row_names = FALSE,
        top_annotation = ha.RNA.L3,
        show_column_names = FALSE,
        col = colorRamp2(c(-2, 0, 2), c("blue", "white", "red")),
        name = "z-score")
```


### Heatmap: Induced & Repressed Genes in IFNG

```{r}
inducedGenes_IFNG_48 <- 
  RNAseq_resExprAcc %>% 
  filter(experimentalCondition == "IFNG_48") %>% 
  filter(Induced) %>% 
  pull(ensembl_gene_id)

repressedGenes_IFNG_48 <- 
  RNAseq_resExprAcc %>% 
  filter(experimentalCondition == "IFNG_48") %>% 
  filter(Repressed) %>% 
  pull(ensembl_gene_id)

Heatmap(RNAseqL3[c(inducedGenes_IFNG_48, repressedGenes_IFNG_48), ], 
        cluster_columns = FALSE, split = rep(c("Induced", "Repressed"), 
                                             c(length(inducedGenes_IFNG_48), length(repressedGenes_IFNG_48))),
        show_row_names = FALSE,
        top_annotation = ha.RNA.L3,
        show_column_names = FALSE,
        col = colorRamp2(c(0, .5, 4), c("blue", "white", "red")),
        name = "log2(fpkm + 1)")

Heatmap(RNAseqL3Z[c(inducedGenes_IFNG_48, repressedGenes_IFNG_48), ], 
        cluster_columns = FALSE, split = rep(c("Induced", "Repressed"), 
                                             c(length(inducedGenes_IFNG_48), length(repressedGenes_IFNG_48))),
        show_row_names = FALSE,
        top_annotation = ha.RNA.L3,
        show_column_names = FALSE,
        col = colorRamp2(c(-2, 0, 2), c("blue", "white", "red")),
        name = "z-score")
```


```{r}
repressedGenesTable <-
  RNAseq_resExprAcc %>%
  filter(experimentalCondition == "IFNG_48") %>%
  filter(Repressed) %>%
  dplyr::select(experimentalCondition, hgnc_symbol, everything())

inducedGenesTable <-
  RNAseq_resExprAcc %>%
  filter(experimentalCondition == "IFNG_48") %>%
  filter(Induced) %>%
  dplyr::select(experimentalCondition, hgnc_symbol, everything())

# outDirPlotsExcel <- "RNAseq/misc/IFNG_inducedRepressedGenes"
# if(!dir.exists(outDirPlotsExcel)) {dir.create(outDirPlotsExcel)}


# write_csv(inducedGenesTable, path = sprintf("%s/MDD_RNAseq_InducedGenes_IFNG48.csv", outDirPlotsExcel))
# write_csv(repressedGenesTable, path = sprintf("%s/MDD_RNAseq_RepressedGenes_IFNG48.csv", outDirPlotsExcel))
# 
# repressedGenesTable
# write.xlsx(inducedGenesTable, 
#            file = sprintf("%s/MDD_RNAseq_InducedRepressedGenes_IFNG48.xlsx", outDirPlotsExcel),
#            sheetName = "Induced", row.names = FALSE)
# write.xlsx(repressedGenesTable, 
#            file = sprintf("%s/MDD_RNAseq_InducedRepressedGenes_IFNG48.xlsx", outDirPlotsExcel),
#            sheetName = "Repressed", row.names = FALSE,
#            append = TRUE)

```

```{r}
aT <-
  getBM(attributes = c("ensembl_gene_id", 
                       "ensembl_transcript_id"),
        mart = mart)

sa.Tornado <- 
  read.csv(ATACseqSAFile, stringsAsFactors = FALSE) %>% 
  # mutate(Peaks = paste0("../", Peaks)) %>% 
  # mutate(bamReads = paste0("../", bamReads)) %>% 
  mutate(bigWig = str_replace(str_remove(bamReads, "ATACseq/Data/Level0/bam/"), ".bam", "_allChr.bigWig"))  %>% 
  left_join(data.frame(bwFile = list.files(bwDir, full.names = TRUE),
                       bigWig = list.files(bwDir), stringsAsFactors = FALSE)) %>% 
  mutate(experimentalCondition = fct_relevel(as.factor(experimentalCondition),
                                             c("ctrl_0", "EGF_24", "EGF_48", "IFNG_24", "IFNG_48"))
  )

txFiltTable <- read.table(transcriptsFile,
                          header = TRUE,
                          stringsAsFactors = FALSE)

prom <- getProm(TxDb.Hsapiens.UCSC.hg38.knownGene, aT)
prom <- prom[prom$tx_name %in% txFiltTable$ensembl_transcript_id]

toPlotInduced <-
  inducedGenesTable %>% 
  dplyr::select(ensembl_gene_id) %>% 
  left_join(at.RNA) %>% 
  distinct(hgnc_symbol, .keep_all = TRUE) %>%     
  mutate(transcriptInRNA = ensembl_gene_id %in% rownames(RNAseqL3),
         promoterInATAC  = ensembl_gene_id %in% prom$ensembl_gene_id)

toPlotRepressed <-
  repressedGenesTable %>% 
  dplyr::select(ensembl_gene_id) %>% 
  left_join(at.RNA) %>% 
  distinct(hgnc_symbol, .keep_all = TRUE) %>%     
  mutate(transcriptInRNA = ensembl_gene_id %in% rownames(RNAseqL3),
         promoterInATAC  = ensembl_gene_id %in% prom$ensembl_gene_id)

tpList <- list("Induced" = toPlotInduced,
               "Repressed"   = toPlotRepressed)

promList <- rep(list(prom), 2)
names(promList) <- names(tpList)

promList <- mapply(function(promoters, moduleGenes) {
  promoters <- promoters[promoters$ensembl_gene_id %in% moduleGenes]
  promoters
  }, 
  prom = promList, 
  moduleGenes = lapply(tpList, function(x) {x <- x %>% pull(ensembl_gene_id)}))

HLA_indices <- lapply(promList, function(promoters) {
  x <- 
    findOverlaps(GRanges("chr6:28510120-33480577"), promoters) %>% 
    data.frame %>% 
    .$subjectHits
    
  x
})

moduleListHLA <-
  mapply(function(moduleDF, promoters, HLA_ind) {
  moduleDF$promoterInHLA <- FALSE
  
  moduleDF$promoterInHLA[match(promoters$ensembl_gene_id[HLA_ind], moduleDF$ensembl_gene_id)] <- TRUE
  
  moduleDF <- filter(moduleDF, !promoterInHLA)
  
  return(moduleDF)
  }, 
  moduleDF = tpList, promoters = promList, HLA_ind = HLA_indices, 
  SIMPLIFY = FALSE)


promNoHLA <- mapply(function(promoters, moduleDF) {
  promoters <- promoters[match(moduleDF %>% 
                                 filter(promoterInATAC) %>% 
                                 filter(!promoterInHLA) %>% 
                                 pull(ensembl_gene_id),
                               promoters$ensembl_gene_id)]
  promoters
  }, 
  promoters = promList,
  moduleDF = moduleListHLA)

geneMats <- lapply(promNoHLA, function(x) {
  x <- ScoreMatrixList(target = sa.Tornado %>%
                         filter(ATACseq_QCpass) %>%
                         filter(ligand %in% c("ctrl", "EGF", "IFNG")) %>%
                         filter(experimentalTimePoint %in% c(0, 48)) %>%
                         arrange(ligand) %>%
                         pull(bwFile),
                       windows = x, 
                       strand.aware = TRUE)
  x
  }
  )

geneMatsNamed <- lapply(geneMats, function(x) {
  names(x) <-
    sa.Tornado %>%
    filter(ATACseq_QCpass) %>%
    filter(ligand %in% c("ctrl", "EGF", "IFNG")) %>%
    filter(experimentalTimePoint %in% c(0, 48)) %>%
    arrange(ligand) %>%
    pull(specimenName)
  x
})

geneMatsCombined <- lapply(geneMatsNamed,
                           combineMat, 
                           combine = list("CTRL" = 1:4,
                                          "EGF" = 5:8,
                                          "IFNG+EGF" = 9:10)
                           )

addAccOrderToModuleDF <- function(gmat, moduleDF) {
  gmat <- lapply(gmat, as, "matrix")
  gmat <- Reduce(cbind, gmat)
  
  sum_acc <- apply(gmat, 1, sum)
  order_acc <- order(sum_acc, decreasing = TRUE)
  
  orderedGenes   <- data.frame(ensembl_gene_id = moduleDF$ensembl_gene_id[order_acc],
                                   rank = seq(1, length(order_acc), 1),
                               stringsAsFactors = FALSE)
  
  moduleDF <- moduleDF %>% 
    left_join(orderedGenes) %>% 
    arrange(rank)
  
  moduleDF
}

moduleListHLAOrdered <- mapply(addAccOrderToModuleDF, 
               gmat = geneMatsCombined, moduleDF = moduleListHLA,
               SIMPLIFY = FALSE)

MDD_RNAseq_Zscore_Medians <-
  RNAseqL3Z %>% 
  rownames_to_column("hgnc_symbol") %>% 
  pivot_longer(-hgnc_symbol, values_to = "z", names_to = "specimenID") %>% 
  left_join(sa.RNA.L3) %>% 
  group_by(hgnc_symbol, experimentalCondition) %>% 
  summarize(medianZ = median(z, na.rm = TRUE)) %>% 
  mutate(medianZ = replace_na(medianZ, 0)) %>% 
  ungroup %>% 
  pivot_wider(names_from = experimentalCondition, values_from = medianZ) %>% 
  column_to_rownames("hgnc_symbol")

sa.RNA.L4 <-
  sa.RNA.L3 %>% 
  distinct(experimentalCondition, .keep_all = TRUE) %>% 
  mutate(experimentalCondition = fct_inorder(as.factor(experimentalCondition))) %>% 
  dplyr::select(experimentalCondition, Ligand, Time) %>% 
  mutate(Ligand = fct_inorder(as.factor(Ligand)),
         Time   = fct_inorder(as.factor(Time)))

MDD_RNAseq_Zscore_Medians <- MDD_RNAseq_Zscore_Medians[, as.character(sa.RNA.L4$experimentalCondition)]

ha.RNA.L4 <- HeatmapAnnotation(
  df = sa.RNA.L4 %>% dplyr::select(-experimentalCondition), 
  col = col
  )



# png(sprintf("%s/MDD_ATACseq_inducedIFNG_tornadoPlot.png", outDirPlots),
#     height = 12, width = 12, units = "in", res = 450)
multiHeatMatrix(geneMatsCombined$Induced,
                winsorize = c(0, 99),
                order = TRUE)
# dev.off()

# pdf(sprintf("%s/MDD_RNAseq_inducedIFNG_heatmap_tornadoOrder.pdf", outDirPlots),
#     height = 10, width = 5)
HeatmapL3C(RNAseqL3Z[moduleListHLAOrdered$Induced$ensembl_gene_id, ],
           cluster_rows = FALSE)
# dev.off()


# pdf(sprintf("%s/MDD_RNAseq_inducedIFNG_heatmapMedian_tornadoOrder.pdf", outDirPlots),
#     height = 10, width = 5)
HeatmapL4C(MDD_RNAseq_Zscore_Medians[moduleListHLAOrdered$Induced$ensembl_gene_id, ],
           cluster_rows = FALSE)
# dev.off()

plotMeta(geneMatsCombined$Induced,
         overlay = TRUE,
         profile.names = c("CTRL", "EGF", "IFNG+EGF"))
```


```{r}
# png(sprintf("%s/MDD_ATACseq_repressedIFNG_tornadoPlot.png", outDirPlots),
#     height = 12, width = 12, units = "in", res = 450)
multiHeatMatrix(geneMatsCombined$Repressed,
                winsorize = c(0, 99),
                order = TRUE)
# dev.off()

# pdf(sprintf("%s/MDD_RNAseq_repressedIFNG_heatmap_tornadoOrder.pdf", outDirPlots),
#     height = 10, width = 5)
HeatmapL3C(RNAseqL3Z[moduleListHLAOrdered$Repressed$ensembl_gene_id, ],
           cluster_rows = FALSE)
# dev.off()
# 
# pdf(sprintf("%s/MDD_RNAseq_repressedIFNG_heatmapMedian_tornadoOrder.pdf", outDirPlots),
#     height = 10, width = 5)
HeatmapL4C(MDD_RNAseq_Zscore_Medians[moduleListHLAOrdered$Repressed$ensembl_gene_id, ],
           cluster_rows = FALSE)
# dev.off()


plotMeta(geneMatsCombined$Repressed, 
         overlay = TRUE, 
         profile.names = c("CTRL", "EGF", "IFNG+EGF"), 
         centralTend = "mean")
```

