---
title: "MDD integrated analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache=TRUE, cache.lazy = FALSE, message=FALSE, warning = FALSE, fig.height=10, fig.width=10)

suppressMessages(library(tidyverse))
suppressMessages(library(ComplexHeatmap))
library(iheatmapr)
library(DT)
library(rrscale)
library(ggbeeswarm)
library(umap)
library(plotly)


pk_preprocess_level3 <- function(df, type){
  #filter to time 0, 24 and 48
  #median summarise
  df_pp <- df %>%
    filter(str_detect(experimentalCondition,"0|24|48")) %>%
    mutate(Type = type) %>%
    group_by(experimentalCondition, feature, Type) %>%
    summarise(value = median(value, na.rm = TRUE)) %>%
    ungroup()
  return(df_pp)
}


select_features <- function(df, var_quantile = 0){
  df_small <- df %>%
    group_by(feature) %>%
    mutate(feature_variance = var(value, na.rm = TRUE)) %>%
    ungroup() %>%
    filter(feature_variance >= quantile(feature_variance,
                                       probs = var_quantile,
                                       na.rm = TRUE)) %>%
    dplyr::select(-feature_variance)
  return(df_small)
}

#' rrscale a numeric vector
rrscaleDiagnostics <- function(x, zeros = .001, ncores = 4){
  x_rr <- as.matrix(x) %>%
    rrscale(zeros = zeros, ncores = ncores)
  return(x_rr)
}

getrrDetails <- function(x){
  T_name <- map(x, function(xx) xx[["T_name"]]) %>%
    unlist
  par_hat <- map(x, function(xx) xx[["par_hat"]]) %>%
    unlist
  df <- tibble(feature = names(x),
               par_hat = par_hat,
               T_name = T_name)
  return(df)
}

#' df a dataframe with feature, value, condition and Type columns. The values must
#' be all numeric and transformed to common scales
get_iheatmap <- function(df, k = 6, ...) {

condition_order <-  c("ctrl_0",paste(rep(c("PBS", "HGF", "OSM", "EGF","BMP2", "IFNG", "TGFB"), each = 2), rep(c(24, 48),  times = 14),   sep = "_") )

  # specimenName_order <-  paste(rep(c("ctrl","PBS", "HGF", "OSM", "EGF","BMP2", "IFNG", "TGFB"), each = 36, times = 3), rep(c(0, 1, 4 ,8, 24, 48), each = 6, times = 3), rep(c("C1", "C2", "C3"), each = 288), rep(c("A", "B", "C", "D", "E", "Med"), times = 144),  sep = "_") %>%
  #   intersect(df$specimenName)
  
df_sp <- df %>%
    dplyr::select(feature, value, experimentalCondition) %>%
    spread(key = experimentalCondition, value = value) %>%
  dplyr::select(feature, condition_order)
  
  df_as_matrix <- df_sp %>%
    dplyr::select(-feature) %>%
    as.matrix()
  rownames(df_as_matrix) <- df_sp$feature
  
  #create top annotations
  ann_nv_pairs <- df %>%
    dplyr::select(feature, Type) %>%
    distinct() %>%
    drop_na()
  
  ann_df <- tibble(feature = rownames(df_as_matrix)) %>%
    left_join(ann_nv_pairs, by = "feature") %>%
    dplyr::select(Type)
  
  #Create the heatmap
  hm <- main_heatmap(data = df_as_matrix,
                     name = "score") %>%
    add_row_labels(font = list(size = 4),
                   side = "right") %>%
    add_col_labels() %>%
    add_row_annotation(ann_df,
                       side = "left",
                       size = 0.05)  %>%
    modify_layout(list(margin = list(r = 120)))
  if(!k==0){
   hm <- hm %>%
     add_row_clustering(name = "Cluster",
                        k = k) %>%
      add_col_summary(size = .1,
                      groups = "Cluster")
  }
  
  return(hm)
}


```


```{r read_data_and_metadata}

md <- read_csv("metadata/MDD_sample_annotations.csv")



#Read in RPPA pathways scores, mean summarise within condition
RPPA_pathways <- read_csv("RPPA/Data/MDD_Pathways_Score.csv") %>%
  rename(specimenID = X1) %>%
  gather(key = feature, value = value, -specimenID) %>%
  inner_join(md, by = "specimenID") %>%
  dplyr::select(experimentalCondition, feature, value) %>%
  pk_preprocess_level3(type =  "RPPA pathway")

#Read in Hallmark pathways that are based on RNAseq data
hallmark_pathways <- read_csv("RNAseq/Data/MDD_RNAseq_HallmarkNES.csv") %>%
  gather(condition, value, -variable) %>%
  rename(feature = variable) %>%
  rename(experimentalCondition = condition) %>%
  mutate(Type = "Hallmark") %>%
  pk_preprocess_level3(type =  "Hallmark")

#Read in RPPA values and merge with metadata
RPPA_values <- read_csv("RPPA/Data/MDD_RPPA_Level3.csv") %>%
  mutate(feature = paste0(antibody,"_RPPA")) %>%
  gather(specimenID, value = value, -feature, -antibody) %>%
  left_join(md, by = "specimenID") %>%
  mutate(value = 2^value) %>%
  pk_preprocess_level3(type = "RPPA")

#Read in and process cycIF values
biomarkers <- c("p21waf1cip1_3_af647cy5_int_mean_nuc",
                "cyclind1_8_af488fitc_int_mean_nuc",
                "ki67_5_efluor570cy3_int_mean_nuc",
                "egfr_2_af488fitc_int_mean_nuc",
                "egfr_2_af488fitc_int_mean_cytoplasm",
                "met_6_af488fitc_int_mean_nuc",
                "met_6_af488fitc_int_mean_cytoplasm",
                "ndg1pt346_3_af488fitc_int_mean_nuc",
                "ndg1pt346_3_af488fitc_int_mean_cytoplasm",
                "s6_6_pecy3_int_mean_nuc",
                "s6_6_pecy3_int_mean_cytoplasm",
                "s6ps235s236_5_af647cy5_int_mean_nuc",
                "s6ps235s236_5_af647cy5_int_mean_cytoplasm",
                "s6ps240244_5_af488fitc_int_mean_nuc",
                "s6ps240244_5_af488fitc_int_mean_cytoplasm",
                "cateninbeta_4_af647cy5_int_mean_nuc",
                "cateninbeta_4_af647cy5_int_mean_cytoplasm",
                "nfkbp65_2_af647cy5_int_mean_nuc",
                "cjun_9_af488fitc_int_mean_nuc",
                "stat1ps727_2_pecy3_int_mean_nuc",
                "stat1alphaisoform_8_af647cy5_int_mean_nuc",
                "pdl1_6_af647cy5_int_mean_cytoplasm",
                "stat3_4_af488fitc_int_mean_nuc",
                "ecadherin_7_af647cy5_int_mean_nuc",
                "ecadherin_7_af647cy5_int_mean_cytoplasm",
                "vimentin_3_af555cy3_int_mean_nuc",
                "vimentin_3_af555cy3_int_mean_cytoplasm",
                "cytokeratin7human_4_af555cy3_int_mean_nuc",
                "cytokeratin7human_4_af555cy3_int_mean_cytoplasm",
                "cytokeratin18_7_af488fitc_int_mean_nuc",
                "cytokeratin18_7_af488fitc_int_mean_cytoplasm",
                "hes1_9_af647cy5_int_mean_nuc",
                "lc3ab_7_af555cy3_txt_standev_cytoplasm")

cycIF_values <- read_csv("cycIF/Data/MDD_cycIF_Level3.csv") %>%
  filter(feature %in% biomarkers) %>%
  gather(specimenID, value = value, -feature) %>%
  spread(key = feature, value = value) %>%
  mutate(egfr_2_af488fitc_int_mean_cell = (egfr_2_af488fitc_int_mean_nuc + egfr_2_af488fitc_int_mean_cytoplasm)/2,
         met_6_af488fitc_int_mean_cell = (met_6_af488fitc_int_mean_nuc + met_6_af488fitc_int_mean_cytoplasm)/2,
         ndg1pt346_3_af488fitc_int_mean_cell = (ndg1pt346_3_af488fitc_int_mean_nuc + ndg1pt346_3_af488fitc_int_mean_cytoplasm)/2,
         s6_6_pecy3_int_mean_cell = (s6_6_pecy3_int_mean_nuc + s6_6_pecy3_int_mean_cytoplasm)/2,
         s6ps235s236_5_af647cy5_cell = (s6ps235s236_5_af647cy5_int_mean_nuc + s6ps235s236_5_af647cy5_int_mean_cytoplasm)/2,
         s6_6_pecy3_int_mean_cell = (s6_6_pecy3_int_mean_nuc + s6_6_pecy3_int_mean_cytoplasm)/2,
         s6ps240244_5_af488fitc_int_mean_cell = (s6ps240244_5_af488fitc_int_mean_nuc + s6ps240244_5_af488fitc_int_mean_cytoplasm)/2,
         cateninbeta_4_af647cy5_int_mean_cell = (cateninbeta_4_af647cy5_int_mean_nuc + cateninbeta_4_af647cy5_int_mean_cytoplasm)/2,
         ecadherin_7_af647cy5_int_mean_cell = (ecadherin_7_af647cy5_int_mean_nuc + ecadherin_7_af647cy5_int_mean_cytoplasm)/2,
         vimentin_3_af555cy3_int_mean_cell = (vimentin_3_af555cy3_int_mean_nuc + vimentin_3_af555cy3_int_mean_cytoplasm)/2,
         cytokeratin7human_4_af555cy3_int_mean_cell = (cytokeratin7human_4_af555cy3_int_mean_nuc + cytokeratin7human_4_af555cy3_int_mean_cytoplasm)/2,
         cytokeratin18_7_af488fitc_int_mean_cell = (cytokeratin18_7_af488fitc_int_mean_nuc + cytokeratin18_7_af488fitc_int_mean_cytoplasm)/2) %>%
  select(-egfr_2_af488fitc_int_mean_nuc,
         -egfr_2_af488fitc_int_mean_cytoplasm,
         -met_6_af488fitc_int_mean_nuc,
         -met_6_af488fitc_int_mean_cytoplasm,
         -ndg1pt346_3_af488fitc_int_mean_nuc,
         -ndg1pt346_3_af488fitc_int_mean_cytoplasm,
         -s6_6_pecy3_int_mean_nuc,
         -s6_6_pecy3_int_mean_cytoplasm,
         -s6ps235s236_5_af647cy5_int_mean_nuc,
         -s6ps235s236_5_af647cy5_int_mean_cytoplasm,
         -s6ps240244_5_af488fitc_int_mean_nuc,
         -s6ps240244_5_af488fitc_int_mean_cytoplasm,
         -cateninbeta_4_af647cy5_int_mean_nuc,
         -cateninbeta_4_af647cy5_int_mean_cytoplasm,
         -ecadherin_7_af647cy5_int_mean_nuc,
         -ecadherin_7_af647cy5_int_mean_cytoplasm,
         -vimentin_3_af555cy3_int_mean_nuc,
         -vimentin_3_af555cy3_int_mean_cytoplasm,
         -cytokeratin7human_4_af555cy3_int_mean_nuc,
         -cytokeratin7human_4_af555cy3_int_mean_cytoplasm,
         -cytokeratin18_7_af488fitc_int_mean_nuc,
         -cytokeratin18_7_af488fitc_int_mean_cytoplasm) %>%
  gather(feature, value, -specimenID) %>%
left_join(md, by = "specimenID") %>%
  mutate(feature = str_remove(feature, "_.*"),
         feature = paste0(feature, "_cycIF")) %>%
  pk_preprocess_level3(type = "cycIF")

#Read in and process GCP values
GCP_values <- read_csv("GCP/Data/MDD_GCP_Level3.csv") %>%
  rename(feature = histone) %>%
  gather(specimenID, value = value, -feature) %>%
  left_join(md, by = "specimenID") %>%
  mutate(value = 2^value) %>%
  filter(!feature == "H3K27ac1K36me0") %>%
  pk_preprocess_level3(type = "GCP")

#Read in RNAseq values
RNA_values <- read_csv("RNAseq/Data/MDD_RNAseq_med.csv") %>%
  drop_na() %>%
  mutate(feature = paste0(feature, "_RNA")) %>%
  rename(experimentalCondition = condition) %>%
  pk_preprocess_level3("RNAseq")

#Read and and process ATACseq metadata
ATACseq_metadata <- read_csv("ATACseq/Metadata/MDD_ATACseq_peakMetadata.csv") %>%
  select(peak, hgnc_symbol, annotation) %>%
  drop_na() %>%
  filter(str_detect(annotation, "Promoter|3' UTR|5' UTR|1st Exon")) %>%
  dplyr::select(-annotation)

#Read in and process ATACseq values
ATACseq_values <- read_csv("ATACseq/Data/MDD_ATACseq_Level3.csv") %>%
  rename(feature = peak) %>%
  gather(specimenID, value = value, -feature) %>%
  left_join(md, by = "specimenID") %>%
  inner_join(ATACseq_metadata, by = c("feature" = "peak")) %>%
    mutate(feature = paste0(hgnc_symbol, "_ATAC")) %>%
  pk_preprocess_level3(type = "ATACseq") %>%
  group_by(experimentalCondition, feature, Type) %>%
  summarise(value = max(value)) %>%
  ungroup()

#Read in and process ATACseq motif values
ATACseq_motif_values <- read_csv("ATACseq/Data/MDD_ATACseq_MotifZscores.csv") %>%
  rename(feature = motif) %>%
  gather(specimenID, value = value, -feature) %>%
  left_join(md, by = "specimenID") %>%
  mutate(feature = str_remove(feature, ".*_"),
         feature = paste0(feature, "_motif")) %>% 
  pk_preprocess_level3(type = "ATACseq_motif")
#Read in and process IF values
IF_values <- read_csv("IF/Data/MCF10A_IF_Ilastik_Image_File.csv") %>%
  select(-ImageNumber, -barcode, -WellIndex, -collection, -ligand) %>%
  gather(feature, value, -specimenName, -time, -replicate) %>%
  mutate(experimentalTimePoint = time,
         experimentalCondition = str_remove(specimenName, "_C[12]_.")) %>%
  filter(!feature =="AreaShape_EulerNumber",
         !str_detect(feature,"CellMask")) %>%
  pk_preprocess_level3(type = "IF") %>%
  drop_na() %>%
  group_by(experimentalCondition, feature, Type) %>%
  summarise(value = median(value, na.rm = TRUE)) %>%
  ungroup()

get_TFs <- function(sheet, dir_path, pattern) {
  #read in and combine MDD TF scores
TFs_input <- dir(path = dir_path,
                 pattern = pattern,
                 full.names = TRUE) %>%
  map(readxl::read_excel, col_types = c("text"), sheet = sheet, .name_repair = "universal") %>%
  bind_rows() %>%
  mutate(Rank = as.numeric(Rank),
         experimentalTime = str_extract(Query.Name, "[24][48]"),
         experimentalTime = as.numeric(experimentalTime),
         ligand = str_remove(Query.Name, "ctrl vs "),
         ligand = str_remove(ligand, ",.*"),
         condition = paste0(ligand, "_", experimentalTime),
         value =(1633-Rank)/1632,
         feature = TF,
         Library_only = str_remove(Library, ",.*"),
         Library_only = str_replace(Library_only, " ", "--"))
}

TFs <- get_TFs(dir_path = "RNAseq/Data/ChEA3_results_MD_OHSU", pattern = "ctrl_vs_.*xlsx", sheet =1)
TFs_details <- map(2:7, get_TFs, dir_path = "RNAseq/Data/ChEA3_results_MD_OHSU", pattern = "ctrl_vs_.*xlsx") %>%
  bind_rows()

TFs_input <- TFs_details %>%
  dplyr::select(Query.Name, TF, Odds.Ratio, FDR, Library_only) %>%
  right_join(TFs, by = c("Query.Name", "TF", "Library_only"))

TFs_values <- TFs_input %>%
  dplyr::select(condition, Odds.Ratio, feature) %>%
  rename(experimentalCondition = condition) %>%
  mutate(feature = paste0(feature, "_TF")) %>%
  rename(value = Odds.Ratio) %>%
  mutate(value = as.numeric(value),
         Type = "ChEA3 TF")

# TFs_down <- read_csv("RNAseq/Data/ChEA3_TFs_down.csv") %>%
#   gather(feature, value, -condition) %>%
#   mutate(Type = "ChEA3 TF Down")

pk_all <- bind_rows(ATACseq_values, ATACseq_motif_values, cycIF_values, GCP_values, hallmark_pathways, IF_values, RNA_values, RPPA_pathways, RPPA_values, TFs_values)

```


```{r prepare_datasets}

ATACseq_variance_probs_thresh <- .995
ATACseq_selected <- select_features(ATACseq_values, ATACseq_variance_probs_thresh)

ATACseq_motif_variance_probs_thresh <- .5
ATACseq_motif_selected <- select_features(ATACseq_motif_values, ATACseq_motif_variance_probs_thresh)

# #Select top TFs of each ligand
odds_ratio_thresh <- 5

TFs_selected_names <- TFs_values %>%
  spread(key = feature, value = value) %>%
  #mutate(Ligand = str_remove(experimentalCondition, "_.*")) %>%
  group_by(experimentalCondition) %>%
  summarise_if(is.numeric, min) %>%
  ungroup %>%
  gather(key = TF, value = Odds.Ratio, -experimentalCondition) %>%
  group_by(experimentalCondition) %>%
  #filter(Odds.Ratio >= odds_ratio_thresh)%>%
  arrange(Odds.Ratio) %>%
  top_n(25) %>%
  ungroup() %>%
  spread(key = TF, value = Odds.Ratio) %>%
  dplyr::select(-experimentalCondition)

TFs_selected <- TFs_values %>%
  spread(key = feature, value = value)  %>%
  dplyr::select(experimentalCondition, colnames(TFs_selected_names)) %>%
  gather(feature, value, -experimentalCondition) %>%
  mutate(Type = "ChEA3 TF") %>%
  drop_na()

#filter RPPA features on variance
RPPA_variance_probs_thresh <- 0
RPPA_selected <- select_features(RPPA_values, RPPA_variance_probs_thresh)

#filter GCP features on variance
GCP_variance_probs_thresh <- 0
GCP_selected <- select_features(GCP_values, GCP_variance_probs_thresh)

#use RNAseq genes filterd on variance within each condition
RNAseq_variance_genes <- read_csv("RNAseq/Data/MDD_RNAseq_geneList.csv") %>%
  mutate(feature= paste0(hgnc_symbol, "_RNA")) %>%
  dplyr::select(-hgnc_symbol)

RNAseq_selected <- RNA_values %>%
  inner_join(RNAseq_variance_genes)


pk_selected <- bind_rows(ATACseq_selected, ATACseq_motif_selected, cycIF_values, GCP_selected, hallmark_pathways, IF_values, RNAseq_selected, RPPA_pathways, RPPA_selected, TFs_selected)

df <- pk_selected %>%
  select(experimentalCondition, feature, value) %>%
#  filter(str_detect(feature,  "BATF")) %>%
  spread(key = feature, value = value) %>%
  select_if(is.numeric)

df_as_list <- matrix(df)
names(df_as_list) <- colnames(df)

rrscale_small <- FALSE
#Conditionally read or compute rrscale results
rrscale_file <- dir("R", pattern = "rrscale.rda", full.names = TRUE)
if(length(rrscale_file)==0){
  if(rrscale_small){
    set.seed(42)
    df_as_list_small <- df_as_list[sample(1:length(df_as_list),size = .1*length(df_as_list),replace = FALSE)]
    all_rr_values <- map(df_as_list_small, rrscaleDiagnostics)
    save(all_rr_values, file = "R/rrscale.rda")
  } else{
    all_rr_values <- map(df_as_list, rrscaleDiagnostics)
    save(all_rr_values, file = "R/rrscale.rda")
  }

} else if(length(rrscale_file)==1){
  load(rrscale_file)
} else {
  stop("There are more than 1 rrscale data files ", rrscale_file)
}

rr_details <- getrrDetails(all_rr_values)

rr_values <- map(all_rr_values, function(xx){
  xx[["RR"]] %>%
  unlist
})  %>%
    bind_cols()

pk_top_rr <- pk_selected %>%
  dplyr::select(-Type) %>%
  spread(key = feature, value = value) %>%
  dplyr::select(experimentalCondition) %>%
  bind_cols(rr_values) %>%
  gather(feature, value, -experimentalCondition) %>%
  mutate(value_rr = signif(value, digits = 3)) %>%
  dplyr::select(-value) %>%
  left_join(pk_selected, by = c("feature", "experimentalCondition")) %>%
  mutate(value = signif(value,digits = 3)) %>%
  filter(!feature %in% c("AreaShape_EulerNumber","H3K27ac1K36me0"))###temp filtering


```

```{r interactiveHeatmaps, fig.height = 12}

df <- pk_top_rr %>%
  select(experimentalCondition, feature, value_rr, Type) %>%
  rename(value = value_rr)
hclust_clust_num <- 8
hm <- get_iheatmap(df, assay_name = "Prior knowledge, rrscale-transformed", k = hclust_clust_num)
hm
df <- pk_top_rr
datatable(df,colnames = c("Condition" = "experimentalCondition",
                          "Feature" = "feature",
                          "Org. value" = "value",
                          "RR value" = "value_rr"))


```

#### Method  

This analysis combines MDD RNAseq, RPPA data with prior knowledge of pathways. It then adds in selected cycIF, RPPA, GCP and IF values. 
This analysis is based on the LINCS Molecular Deep Dive data files on Synapse at  https://www.synapse.org/#!Synapse:syn2862345/wiki/588244.  

RPPA pathways   
Use the MDACC algorithm and OHSU's antibody-to-pathway assignments to create pathway scores for each , 0, 24 and 48 hour sample.  
Median summarize the pathway scores to the treatment level.  
There are `r length(unique(RPPA_pathways$feature))` RPPA pathways in the heatmap.    

Hallmark pathways  
Hallmark pathways are based on the RNAseq data.  
There are `r length(unique(hallmark_pathways$feature))` Hallmark pathways in the heatmap.    

cycIF Values  
Select `r length(unique(cycIF_values$feature))` cycIF features based on their biological relevance  
Median summarize the values across the replicates  

RPPA Values  
Median summarize the values across the replicates  
Select the RPPA antibodies with a variance above the `r RPPA_variance_probs_thresh*100`th percentile. 
This yields `r length(unique(RPPA_selected$feature))` high variance antibodies in the heatmap.  

GCP Values  
Median summarize the values across the replicates  
Select the GCP histones with a variance above the `r GCP_variance_probs_thresh*100`th percentile. This yields `r length(unique(GCP_selected$feature))` high variance histones in the heatmap. 

RNAseq Values  
A gene list representative of the largest gene expression changes across all ligand treatments was produced from differential expression analyses of each experimental condition compared to the control condition. For each experimental condition, a differential expression analysis was performed using the R package DESeq2 (1.24.0), with shrunken log2 fold change estimates calculated using the apeglm method. Each list was filtered to include only genes with an adjusted p value < 0.05, then ranked by the absolute value of their log2 fold change estimate. The 25 most differentially expressed genes per condition were selected. Redundant genes were removed, yielding a list of `r length(unique(RNAseq_selected$feature))` unique genes included in the heatmap.

ATACseq Values  
Median summarize the values across the replicates  
Select the genes with a variance above the `r ATACseq_variance_probs_thresh*100`th percentile. This yields `r length(unique(ATACseq_selected$feature))` high variance genes in the heatmap.


ATACseq motifs
Read in supplied ATACseq motif values. 
There are `r length(unique(ATACseq_motif_values$feature))` ATACseq motifs in the heatmap.

IF Values  
Filter out the CellMask intensities and use the remaining `r length(unique(IF_values$feature))` IF features  
Median summarize the values across the replicates  

ChEA3 TF Values  
There are `r length(unique(TFs_selected$feature))` ChEA3 TFs in the heatmap.    

Clustering  
The hierarchical clustering uses the complete linkage method to find similar clusters. The number of clusters as shown in the side annotation bar is arbitrarily chosen as `r hclust_clust_num`. The scatter plot above the heatmap shows the mean value for each cluster within each condition.  

RRscale  
Use rrscale to transform each feature in the heatmap before clustering. This method performs a box-cox transformation on each feature, zscores the results then eliminates outliers with absolute values greater than 4.

```{r plotClusters, fig.height = 14}
#Start with an existing iheatmap object
#Go through each cluster, one at a time
#Filter the original data by the row names
#Create a new iheatmap without additional row clusters
# res <- lapply(unique(hm@plots@listData$Cluster@data)[5], function(cl){
#   cl_rows <- hm@plots@listData$Cluster@data[hm@plots@listData$Cluster@data== cl]
#   df <- pk_top_rr %>% 
#     filter(feature %in% names(cl_rows)) %>%
#   select(condition, feature, value_rr, Type) %>%
#   rename(value = value_rr)
# 
# hm_one_cl <- get_iheatmap(df, assay_name = "Prior knowledge, rrscale-transformed", k = 0)
# })

cl <- 6
  cl_rows <- hm@plots@listData$Cluster@data[hm@plots@listData$Cluster@data== cl]
  df <- pk_top_rr %>% 
    filter(feature %in% names(cl_rows)) %>%
  select(experimentalCondition, feature, value_rr, Type) %>%
  rename(value = value_rr)

hm_one_cl <- get_iheatmap(df, assay_name = "Prior knowledge, rrscale-transformed", k = 0)

hm_one_cl



```
The above heatmap is cluster `r cl` from the main heatmap.  

```{r plot_UMAP, eval = FALSE}

df_sp <-  pk_top_rr %>%
  dplyr::select(-value) %>%
  rename(value = value_rr) %>%
  spread(key = experimentalCondition, value = value) %>%
  filter(!is.na(Type))

df_mat <- df_sp %>%
  dplyr::select_if(is.numeric) %>%
  #t() %>%
  as.matrix()
rownames(df_mat) <- df_sp$feature

df_cluster <- tibble(Cluster = hm@plots@listData$Cluster@data,
                  feature = names(hm@plots@listData$Cluster@data))
df_UMAP <- umap(df_mat, na.rm = TRUE)$layout %>%
  data.frame() %>%
  rename(UMAP_1 = X1,
         UMAP_2 = X2) %>%
  mutate(feature = df_sp$feature,
         Type = df_sp$Type) %>%
  left_join(df_cluster)

set.seed(42)
p <- ggplot(df_UMAP, aes(x = UMAP_1,
                         y= UMAP_2,
                         colour = feature,
                         fill = Type)) +
  geom_point(alpha = .5) +
  geom_point(shape = 21,  alpha = .8, stroke = .01) +
  guides(colour = FALSE,
         fill = FALSE) +
  labs(title = paste("UMAP embedding of MDD integrated data")) +
  theme_bw()

ggplotly(p)

p <- ggplot(df_UMAP, aes(x = UMAP_1,
                         y= UMAP_2,
                         colour = as.factor(Cluster))) +
  geom_point(alpha = .5) +
  labs(title = paste("UMAP embedding of MDD integrated data")) +
  theme_bw()
p


# Each of the `r nrow(df_UMAP)` dots represents an analyte with its `r ncol(df_sp)-2` condition values reduced to 2 UMAP values. The first interactive map enables identifying the analyzte. The second static UMAP shows the cluster assignments made in the hierarchical clustering of the main interactive heatmap.  

```



```{r correlation, fig.width=12, eval = TRUE}
df_sp <-  pk_top_rr %>%
  dplyr::select(-value) %>%
  rename(value = value_rr) %>%
  spread(key = experimentalCondition, value = value) %>%
  filter(!is.na(Type))

df_mat <- df_sp %>%
  dplyr::select_if(is.numeric) %>%
  #t() %>%
  as.matrix()
rownames(df_mat) <- df_sp$feature

  title_suffix <- ""

    row_nums <- which(rownames(df_mat) %in% c(hallmark_pathways$feature, RPPA_pathways$feature))
    ra <- rowAnnotation(foo = anno_mark(at = c(row_nums), labels = rownames(df_mat)[row_nums]))
  #Create the heatmap
  shm <- Heatmap(matrix = df_mat %>%
#            t %>%
            #scale(scale = FALSE) %>%
            t %>%
            cor(use = "complete.obs",method = "spearman"),
          name = "correlation",
 #         column_title = paste0("Correlation of ",assay_name, title_suffix),
 #         top_annotation = create_top_annotations(df, md),
 #         left_annotation = haRow,
          right_annotation = ra,
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = TRUE,
          cluster_columns = TRUE,
          col = circlize::colorRamp2(c(-1, 0, 1), c("blue", "white", "red")),
          na_col = "grey")
shm
```

```{r fig.width=8}

 shm_order <- row_order(shm)
#Create the heatmap
  cor_hm <- main_heatmap(data = df_mat %>%
            t %>%
            cor(use = "complete.obs",method = "spearman"),
name = "score",
row_order = rev(shm_order),
col_order = shm_order) %>%
    add_row_labels(font = list(size = 4),
                   side = "right") %>%
    #     add_row_clustering(method = "hclust") %>%
    # #add_col_labels() %>%
    #     add_col_clustering(method = "hclust") %>%

 #   add_row_annotation(ann_df,
 #                      side = "left",
 #                      size = 0.05)  %>%
    modify_layout(list(margin = list(r = 120)))
cor_hm
   

```

Each row and column in this heatmap of correlations is an analyte from the main heatmap. The callouts on the right are the `r length(unique(hallmark_pathways$feature))` hallmark pathways identified using the RNAseq data.  




```{r rrscale_disgnostics, fig.height=5, eval=FALSE}

df <- pk_top %>%
  select(feature, Type) %>%
  distinct()%>%
  right_join(rr_details, by = "feature")

p <- ggplot(df, aes(factor(Type), par_hat, colour = T_name)) +
  geom_boxplot() +
  labs(title = "Box-Cox transformations and lambdas",
       colour = "Transformation",
       y = "lambda",
       x = "assay")
p

```

```{r beforeAfterTransforms, fig.height=4, eval = FALSE}
set.seed(42)
select_features <- pk_top_rr %>%
  group_by(Type) %>%
  select(feature) %>%
  distinct() %>%
  sample_n(9) %>%
  ungroup() %>%
  unlist
  
df <- pk_top_rr %>%
  filter(feature %in% select_features) %>%
  left_join(rr_details, by = "feature")

for(oneType in unique(df$Type)){
  dft <- df %>%
    filter(Type == oneType)

p <- ggplot(dft, aes(x =value)) +
  geom_density() +
  facet_grid(~feature,scales = "free") +
  labs(title = paste(oneType, "not-transformed"))
print(p)

p <- ggplot(dft, aes(x = value_rr, label = par_hat)) +
  geom_density() +
  facet_grid(~feature,scales = "free")+
  labs(title = paste(oneType, "rrscale-transformed"))
print(p)

dftd <- dft %>%
  group_by(feature) %>%
  mutate(value = median(value),
         value_rr = median(value_rr, na.rm = TRUE),
         par_hat = sprintf("%3.2f", par_hat)) %>%
  ungroup() %>%
  select(-condition) %>%
  distinct() 
p <- ggplot(dft, aes(x = value, y = value_rr)) +
  geom_line() +
  facet_grid(~feature,scales = "free")+
  labs(title = paste(oneType, "rrscale-transformed"))
p <- p + geom_label(data=dftd, aes(label = as.character(par_hat)))
print(p)
}

boxcox <- function(x,lambda) {
  if(lambda == 0) {
    y = log(x)
  } else {
    y <- (sign(x) * abs(x)^lambda - 1)/lambda
  }
  return(y)
}

bc <- tibble(x = seq(-3,3, by = .01)) %>%
  mutate(minus_2 = boxcox(x, lambda = -2),
         minus_1 = boxcox(x, lambda = -1),
         minus_.5 = boxcox(x, lambda = -.5),
         pospt_5 = boxcox(x, lambda = .5),
         pos_1 = boxcox(x, lambda = 1),
         pos_2 = boxcox(x, lambda = 2)) %>%
  gather(key = lambda, value, -x)

p <- ggplot(bc, aes(x, value, color = lambda)) +
  geom_line() +
  coord_cartesian(y = c(-10,10))
p

```
