---
title: "MDD integrated analysis heatmap clusters"
output: html_document
---

```{r setup, include=FALSE}
####ToDOos
#change distance calculation in clustering
#do cluster analysis to set cluster number
#fix Type colors and use in bar chart

knitr::opts_chunk$set(echo = FALSE, cache=TRUE, cache.lazy = FALSE, message=FALSE, warning = FALSE, fig.height=10, fig.width=12)

suppressMessages(library(tidyverse))
suppressMessages(library(ComplexHeatmap))
library(iheatmapr)
library(DT)
library(rrscale)
library(ggbeeswarm)
library(umap)
library(plotly)
library(cluster)
library(fpc)
```

```{r functions}

#prepare a datamatrix from a dataframe of data and metadta
prep_hm_matrix <- function(df){
  
  condition_order <-  c("ctrl_0",paste(rep(c("PBS", "HGF", "OSM", "EGF","BMP2", "IFNG", "TGFB"), each = 2), rep(c(24, 48),  times = 14),   sep = "_") )
  
  condition_order <- condition_order[condition_order %in% df$experimentalCondition]
  df_sp <- df %>%
    dplyr::select(feature, value, experimentalCondition) %>%
    spread(key = experimentalCondition, value = value, fill = median(df$value, na.rm = TRUE)) %>%
    dplyr::select(feature, condition_order)
 #   drop_na()
  
  df_as_matrix <- df_sp %>%
    dplyr::select(-feature) %>%
    as.matrix()
  rownames(df_as_matrix) <- df_sp$feature
  return(df_as_matrix)
}

prep_hm_annotations <- function(df){
  df_as_matrix <- prep_hm_matrix(df)
  #create top annotations
  ann_nv_pairs <- df %>%
    dplyr::select(feature, Type) %>%
    distinct()
#    drop_na()
  
  ann_df <- tibble(feature = rownames(df_as_matrix)) %>%
    left_join(ann_nv_pairs, by = "feature") %>%
    dplyr::select(Type)
  return(ann_df)
}

get_iheatmap <- function(df_as_matrix, ...) {
  #Create the heatmap
  hm <- main_heatmap(data = df_as_matrix,
                     name = "score")
}
format_hm <- function(hm, k = 6, cluster_method = "hclust",...){
  hm_mod<- hm %>%
    add_col_labels() %>%
    add_row_annotation(ann_df,
                       side = "left",
                       size = 0.05)  %>%
    modify_layout(list(margin = list(r = 120)))
  if(!k==0){
    hm_mod <- hm_mod %>%
      add_row_clustering(name = paste0("Cluster(",cluster_method,")"),
                         k = k,
                         method = cluster_method,
      ) 
  }
  return(hm_mod)
}

```


```{r read_data_and_metadata}

# if(file.exists("R/all_pk_data.rda")){
#     load("R/all_pk_data.rda")
# }
```


```{r prepare_datasets}
ATACseq_variance_probs_thresh <- .995
ATACseq_motif_variance_probs_thresh <- .5
GCP_variance_probs_thresh <- 0

RPPA_variance_probs_thresh <- 0
odds_ratio_thresh <- 5

if(file.exists("R/selected_pk_data.rda")){
    load("R/selected_pk_data.rda")
}

zscore_cutoff <- Inf
if(file.exists("R/rrscale_assay.rda")){
  load("R/rrscale_assay.rda")
} 


```

## {.tabset .tabset-fade}

### heatmap

```{r interactiveHeatmaps, fig.height = 8}

df_as_matrix <-  prep_hm_matrix(pk_selected_rr) %>%
  t %>%
  as.matrix() %>%
  scale %>%
  t %>%
  signif(digits = 4)

ann_df <- prep_hm_annotations(pk_selected_rr)

clust_num <- 11
cluster_method = "kmeans"
set.seed(42)
hm <- get_iheatmap(df_as_matrix, assay_name = "Prior knowledge, rrscale-transformed") %>%
  format_hm(k = clust_num, cluster_method = cluster_method) 
hm

hm_data <- tibble(feature = names(hm@plots@listData$`Cluster(kmeans)`@data),
                  Cluster = hm@plots@listData$`Cluster(kmeans)`@data,
                  Type = hm@plots@listData$Type@text)

pk_selected_rr_sp <- df_as_matrix %>%
  data.frame() %>%
  rownames_to_column("feature") %>%
  left_join(hm_data, by = "feature")

```




```{r staticHeatmap, fig.width=14, eval = TRUE}
#create a second, static heatmap to have larger labels
y_order <- hm@yaxes@listData[["y"]]@order %>%
  rev()

row_nums <- which(rownames(df_as_matrix) %in% c(hallmark_pathways$feature, RPPA_pathways$feature))
ra <- rowAnnotation(foo = anno_mark(at = c(row_nums), labels = rownames(df_as_matrix)[row_nums]))
#Create the heatmap
shm <- Heatmap(matrix = df_as_matrix,
               name = "",
               #         column_title = paste0("Correlation of ",assay_name, title_suffix),
               #         top_annotation = create_top_annotations(df, md),
               #         left_annotation = haRow,
               right_annotation = ra,
               show_row_names = FALSE,
               row_order = y_order,
               show_column_names = TRUE,
               cluster_rows = FALSE,
               cluster_columns = FALSE,
               #col = circlize::colorRamp2(c(-1, 0, 1), c("blue", "white", "red")),
               na_col = "grey")
shm

```

## UMAP

```{r UMAPAnalysis, fig.width=5.75, fig.height=4.75}

#Generate a 2d scatterplot of the UMAP dimensions and color by clusters, assay type

#Create annotation values

hm_data_matrix <- pk_selected_rr_sp %>%
  select(-feature, -Cluster, -Type) %>%
  as.matrix()
rownames(hm_data_matrix) <- pk_selected_rr_sp$feature

df_UMAP <- umap(hm_data_matrix, na.rm = TRUE)$layout %>%
  data.frame(feature = rownames(hm_data_matrix)) %>%
  rename(UMAP_1 = X1,
         UMAP_2 = X2) %>%
 left_join(pk_selected_rr_sp, by = "feature") 

p <- ggplot(df_UMAP, aes(x = UMAP_1,
                         y = UMAP_2,
                         colour = factor(Cluster))) +
  geom_point(size = .8, alpha = .8) +
  labs(title = paste("UMAP embedding colored by cluster"),
       colour = "Cluster") +
  theme_bw() +
  theme(axis.title = element_blank())
p
```


```{r UMAPAnalysis2, fig.width=6.5, fig.height=4.5}
p <- ggplot(df_UMAP, aes(x = UMAP_1,
                         y = UMAP_2,
                         colour = factor(Type))) +
  geom_point(size = .8, alpha = .8) +
  labs(title = paste("UMAP embedding colored by assay"),
       colour = "Assay") +
  theme_bw() +
  theme(axis.title = element_blank())
p
```

```{r UMAPbyCondition, fig.width=2.5, fig.height=2.5}
for(condition in unique(pk_selected_rr$experimentalCondition)){
  p <- ggplot(df_UMAP, aes_string(x = "UMAP_1",
                         y = "UMAP_2",
                         colour = condition)) +
  geom_point(size = .4) +
    annotate("text",x = -5, y = -6, label = condition)+
  scale_color_gradient2(low = "blue4", mid = "white",
  high = "red", midpoint = 0,) +
  labs()+
    guides(colour = FALSE)+
  theme_bw()+
    theme(text = element_text(size = 1),
          axis.title = element_blank())
print(p)
}

pk_selected_rr_sp %>%
  filter(Type %in% c("Hallmark", "RPPA pathway")) %>%
  select(feature, Cluster, Type) %>%
  datatable(caption = 'Hallmark and RPPA pathway cluster assignments')

```

```{r clusterAnalysis, fig.width=6.5, fig.height=4.5}
#create a bar chart of the feature types
df <- pk_selected_rr_sp %>%
  select(Cluster, Type)

p <- ggplot(df, aes(x = Cluster, fill = Type)) +
  geom_bar() +
  labs(title = paste("Type distribution of the",dim(df)[1], "features by cluster"),
       x ="Cluster") +
  theme_bw()
p

```


### method  

This dataset combines MCF10A Molecular Deep Dive (MDD) RNAseq, Reverse Phase Protein Array(RPPA), Global Chromatin Profiling (GCP) histone modifications, immunofluorescence (IF), cyclic immunofluorescence (cycIF) and ATACseq chromatin configuration data with prior knowledge of signaling pathways, DNA motifs and transcription factors. All data is collected at 0, 24 and 48 hours. Most of the data is median summarized from three replicate experiments run in sequential weeks. It is then rrscale transformed and then zscored by feature.   

RRscale  
Use rrscale https://CRAN.R-project.org/package=rrscale to transform features in the heatmap before clustering.  This method performs a box-cox gaussianizing transformation on each feature, zscores the results then eliminates outliers with absolute values greater than `r zscore_cutoff`. All values from each assay was rrscaled as a matrix with features in the columns and conditions in the rows. Because of the differences in scales in the morphology, spatial and statistical values in the IF dataset, only the intensity values are included.      

This analysis is based on the LINCS Molecular Deep Dive data files on Synapse at  https://www.synapse.org/#!Synapse:syn2862345/wiki/588244.  


RNAseq Values  
A gene list representative of the largest gene expression changes across all ligand treatments was produced from differential expression analyses of each experimental condition compared to the control condition. For each experimental condition, a differential expression analysis was performed using the R package DESeq2 (1.24.0), with shrunken log2 fold change estimates calculated using the apeglm method. Each list was filtered to include only genes with an adjusted p value < 0.05, then ranked by the absolute value of their log2 fold change estimate. The 25 most differentially expressed genes per condition were selected. Redundant genes were removed, yielding a list of `r length(unique(RNAseq_selected$feature))` unique genes included in the heatmap.  

RPPA Values  
Median summarize the values across the replicates  
Select the RPPA antibodies with a variance above the `r RPPA_variance_probs_thresh*100`th percentile. 
This yields `r length(unique(RPPA_selected$feature))` high variance antibodies in the heatmap.  

GCP Values  
Median summarize the values across the replicates  
Select the GCP histones with a variance above the `r GCP_variance_probs_thresh*100`th percentile. This yields `r length(unique(GCP_selected$feature))` high variance histones in the heatmap.  

IF Values  
Filter out the CellMask intensities and use the remaining `r length(unique(IF_values$feature))` IF features  
Median summarize the values across the replicates  

cycIF Values  
Select `r length(unique(cycIF_values$feature))` cycIF features based on their biological relevance  
Median summarize the values across the replicates  

ATACseq Values  
Median summarize the values across the replicates  
Select the genes with a variance above the `r ATACseq_variance_probs_thresh*100`th percentile. This yields `r length(unique(ATACseq_selected$feature))` high variance genes in the heatmap.  

RPPA pathways   
Use the MDACC algorithm and OHSU's antibody-to-pathway assignments to create pathway scores for each , 0, 24 and 48 hour sample.  
Median summarize the pathway scores to the treatment level.  
There are `r length(unique(RPPA_pathways$feature))` RPPA pathways in the heatmap.    

Hallmark pathways    
Hallmark pathways are based on the RNAseq data.  
There are `r length(unique(hallmark_pathways$feature))` Hallmark pathways in the heatmap.    

ATACseq motifs  
Read in supplied ATACseq motif values. 
There are `r length(unique(ATACseq_motif_selected$feature))` ATACseq motifs in the heatmap.  

ChEA3 TF Values  
There are `r length(unique(TFs_selected$feature))` ChEA3 TFs in the heatmap.    

Clustering  
The hierarchical clustering uses the complete linkage method to find similar clusters. The number of clusters as shown in the side annotation bar. It is chosen as `r clust_num` based on gap statistic analysis. The scatter plot above the heatmap shows the mean value for each cluster within each condition.  


```{r featureTypes, fig.width=8, fig.height=4}
#create a bar chart of the feature types
df <- pk_selected_rr %>%
  drop_na() %>%
  select(feature, Type) %>%
  distinct()
p <- ggplot(df, aes(x = Type, fill = Type)) +
  geom_bar() +
  labs(title = paste("Type distribution of the",length(unique(df$feature)), "features in the MDD integrated dataset"),
       x ="Feature type") +
  theme(axis.text = element_text(angle = 90))
p

```
