---
title: "Upset analysis of the integrated MDD dataset"
output: html_document
---

```{r setup, include=FALSE}
####ToDOos
#change distance calculation in clustering
#do cluster analysis to set cluster number
#fix Type colors and use in bar chart

knitr::opts_chunk$set(echo = FALSE, cache=TRUE, cache.lazy = FALSE, message=FALSE, warning = FALSE, fig.height=10, fig.width=12)

suppressMessages(library(tidyverse))
library(UpSetR)
library(DT)
```


```{r read_data_and_metadata}

load("R/rrscale_assay.rda")

df_sp <- pk_selected_rr %>%
  spread(experimentalCondition, value) 

cor_mat <- df_sp %>%
  select_if(is.numeric) %>%
            as.matrix() %>%
            t %>%
            cor(use = "complete.obs", method = "spearman")

rownames(cor_mat) <- df_sp$feature
colnames(cor_mat) <- df_sp$feature

cor_df_all <- cor_mat %>%
  as_tibble() %>%
  mutate(feature = df_sp$feature) %>%
  select(feature, everything()) %>%
  gather(feature2, value, -feature)

```


<bksp>  


## {.tabset .tabset-fade}




### Condition Values{.tabset .tabset-fade}

#### Up/down

```{r ligandSets}

zscore_thresh <- 1

# define members of ligad sets as having any reading above a threshold
#then do upset analysis on those ligand sets
set_membership <- function(x, zscore_thresh = 1.5){
  res <- numeric(length(x))
  res[abs(x) > zscore_thresh] <- 1
  return(res)
}
df <- pk_selected_rr %>%
  mutate(value = set_membership(value, zscore_thresh = zscore_thresh)) %>%
  spread(experimentalCondition, value, fill = 0)  %>%
  as.data.frame()
  

upset(df, nsets = 15, number.angles = 30, point.size = 3.5, line.size = 2, nintersects = 40, order = "freq",
    mainbar.y.label = "Condition sets Intersections", sets.x.label = "Total Features Per Condition", 
    text.scale = c(1.3, 1.3, 1, 1, 1, 0.75))

```


Features are a member of a condition set if the rrscale absolute value is greater than `r zscore_thresh`.  

#### Up


```{r ligandSets_up}

# define members of ligad sets as having any reading above a threshold
#then do upset analysis on those ligand sets
set_membership_up <- function(x, zscore_thresh = 1.5){
  res <- numeric(length(x))
  res[x > zscore_thresh] <- 1
  return(res)
}
df <- pk_selected_rr %>%
  mutate(value = set_membership_up(value, zscore_thresh = zscore_thresh)) %>%
  spread(experimentalCondition, value, fill = 0)  %>%
  as.data.frame()
  

upset(df, nsets = 15, number.angles = 30, point.size = 3.5, line.size = 2, nintersects = 40, order = "freq",
    mainbar.y.label = "Condition sets Intersections", sets.x.label = "Total Features Per Condition", 
    text.scale = c(1.3, 1.3, 1, 1, 1, 0.75))

```

Features are a member of a condition set if the rrscale value is greater than `r zscore_thresh`.  

#### Down


```{r ligandSets_down}

# define members of ligad sets as having any reading below a threshold
#then do upset analysis on those ligand sets
set_membership_down <- function(x, zscore_thresh = 1.5){
  res <- numeric(length(x))
  res[x < -zscore_thresh] <- 1
  return(res)
}
df <- pk_selected_rr %>%
  mutate(value = set_membership_down(value, zscore_thresh = zscore_thresh)) %>%
  spread(experimentalCondition, value, fill = 0)  %>%
  as.data.frame()
  

upset(df, nsets = 15, number.angles = 30, point.size = 3.5, line.size = 2, nintersects = 40, order = "freq",
    mainbar.y.label = "Condition sets Intersections", sets.x.label = "Total Features Per Condition", 
    text.scale = c(1.3, 1.3, 1, 1, 1, 0.75))


```

Features are a member of a condition set if the rrscale value is less than -`r zscore_thresh`.

###  

### RPPA_24 Values{.tabset .tabset-fade}

#### Up/Down


```{r RPPA_24HligandSets_updown}

zscore_thresh <- 1

# define members of ligand sets as having any reading above a threshold
#then do upset analysis on those ligand sets
set_membership <- function(x, zscore_thresh = 1.5){
  res <- numeric(length(x))
  res[abs(x) > zscore_thresh] <- 1
  return(res)
}
df <- pk_selected_rr %>%
  filter(Type == "RPPA") %>%
  filter(str_detect(experimentalCondition, "_24")) %>%
  mutate(value = set_membership(value, zscore_thresh = zscore_thresh)) %>%
  spread(experimentalCondition, value, fill = 0)  %>%
  as.data.frame()
  

upset(df, nsets = 15, number.angles = 30, point.size = 3.5, line.size = 2, nintersects = 40, order = "freq",
    mainbar.y.label = "Condition sets Intersections", sets.x.label = "Total Features Per Condition", 
    text.scale = c(1.3, 1.3, 1, 1, 1, 0.75))

```


Features are a member of a condition set in the RPPA 24 hour data if their absolute value is greater than `r zscore_thresh`.  

#### Up


```{r RPPA_24HligandSets_up}

# define members of ligad sets as having any reading above a threshold
#then do upset analysis on those ligand sets
set_membership_up <- function(x, zscore_thresh = 1.5){
  res <- numeric(length(x))
  res[x > zscore_thresh] <- 1
  return(res)
}
df <- pk_selected_rr %>%
  filter(Type == "RPPA") %>%
  filter(str_detect(experimentalCondition, "_24")) %>%
  mutate(value = set_membership_up(value, zscore_thresh = zscore_thresh)) %>%
  spread(experimentalCondition, value, fill = 0)  %>%
  as.data.frame()
  

upset(df, nsets = 15, number.angles = 30, point.size = 3.5, line.size = 2, nintersects = 40, order = "freq",
    mainbar.y.label = "Condition sets Intersections", sets.x.label = "Total Features Per Condition", 
    text.scale = c(1.3, 1.3, 1, 1, 1, 0.75))

```

Features are a member of a condition set in the RPPA 24 hour data if the value is greater than `r zscore_thresh`.  

#### Down

```{r RPPA_24HligandSets_down}

# define members of ligad sets as having any reading above a threshold
#then do upset analysis on those ligand sets
set_membership_down <- function(x, zscore_thresh = 1.5){
  res <- numeric(length(x))
  res[x < zscore_thresh] <- 1
  return(res)
}
df <- pk_selected_rr %>%
  filter(Type == "RPPA") %>%
  filter(str_detect(experimentalCondition, "_24")) %>%
  mutate(value = set_membership_down(value, zscore_thresh = zscore_thresh)) %>%
  spread(experimentalCondition, value, fill = 0)  %>%
  as.data.frame()
  

upset(df, nsets = 15, number.angles = 30, point.size = 3.5, line.size = 2, nintersects = 40, order = "freq",
    mainbar.y.label = "Condition sets Intersections", sets.x.label = "Total Features Per Condition", 
    text.scale = c(1.3, 1.3, 1, 1, 1, 0.75))

```


Features are a member of a condition set in the RPPA 24 hour data if the value is less than -`r zscore_thresh`.  


###

### Correlations{.tabset .tabset-fade}


#### RPPA Pathways

```{r upsetPlots}
cor_thresh <- .7

#define sets as the RPPA pathways
RPPA_pathways <- pk_selected_rr %>%
  filter(str_detect(feature, "Pathway_Score")) %>%
  select(feature) %>%
  distinct

#get the features with high correlations to the RPPA pathways
df <- cor_df_all %>%
  right_join(RPPA_pathways) %>%
  filter(value > cor_thresh) %>%
  mutate(value = 1L,
         Type = str_remove(feature2, ".*_")) %>%
  spread(key = feature, value = value, fill = 0L) %>%
  as.data.frame()
  
upset(df, nsets = 9, number.angles = 30, point.size = 3.5, line.size = 2, nintersects = 40, order = "freq",
    mainbar.y.label = "RPPA Pathway Intersections", sets.x.label = "Features Per Pathway", 
    text.scale = c(1.3, 1.3, 1, 1, 2, 0.75))
```


Features are a member of a condition set when the Spearman correlation rho value to an RPPA pathway is greater than `r cor_thresh`.  


#### RNA Hallmarks

```{r}

#define sets as the RNA hallmarks
RNA_hallmarks <- pk_selected_rr %>%
  filter(str_detect(Type, "Hallmark")) %>%
  select(feature) %>%
  distinct

#get the features with high correlations to the RPPA pathways
df <- cor_df_all %>%
  right_join(RNA_hallmarks) %>%
  filter(value > cor_thresh) %>%
  mutate(value = 1L,
         Type = str_remove(feature2, ".*_")) %>%
  spread(key = feature, value = value, fill = 0L) %>%
  as.data.frame()
  

upset(df, nsets = 20, number.angles = 30, point.size = 3.5, line.size = 2, nintersects = 40, order = "freq",
    mainbar.y.label = "RNAseq Hallmark Intersections", sets.x.label = "Features Per Hallmark", 
    text.scale = c(1.3, 1.3, 1, 1, 1, 0.75))
```


Features are a member of a condition set when the Spearman correlation rho value to an RNA Hallmark is greater than `r cor_thresh`.  


#### Correlation Table

```{r}
df <- cor_df_all %>%
  filter(value > cor_thresh) %>%
  mutate(value = signif(value, digits = 3))
datatable(df)

```

####

###

### Method  

This dataset combines MCF10A Molecular Deep Dive (MDD) RNAseq, Reverse Phase Protein Array(RPPA), Global Chromatin Profiling (GCP) histone modifications, immunofluorescence (IF), cyclic immunofluorescence (cycIF) and ATACseq chromatin configuration data with prior knowledge of signaling pathways, DNA motifs and transcription factors. All data is collected at 0, 24 and 48 hours. Most of the data is median summarized from three replicate experiments run in sequential weeks. It is then rrscale transformed and then z scaled by assay.     

RRscale  
Use rrscale https://CRAN.R-project.org/package=rrscale to transform features in the heatmap before clustering.  This method performs a box-cox gaussianizing transformation on each feature and then standardizes them. All values from each assay were rrscaled as a matrix with features in the columns and conditions in the rows. Because of the differences in scales in the morphology, spatial and statistical values in the IF dataset, only the intensity values are included.      

This analysis is based on the LINCS Molecular Deep Dive data files on Synapse at  https://www.synapse.org/#!Synapse:syn2862345/wiki/588244.  

The selected dataset is used in this analysis. The criteria are explained in a companion  report. 


```{r featureTypes, fig.width=8, fig.height=4}
#create a bar chart of the feature types
df <- pk_selected_rr %>%
  drop_na() %>%
  select(feature, Type) %>%
  distinct()
p <- ggplot(df, aes(x = Type, fill = Type)) +
  geom_bar() +
  labs(title = paste("Type distribution of the",length(unique(df$feature)), "features in the MDD integrated dataset"),
       x ="Feature type") +
  theme(axis.text = element_text(angle = 90))
p

```


###

##