---
title: "MDD standard figures"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache=TRUE,  message=FALSE, warning = FALSE, fig.height=10, fig.width=10)

suppressMessages(library(tidyverse))
suppressMessages(library(ComplexHeatmap))
library(umap)
library(biomaRt)

```


## Overview{.tabset .tabset-fade}
These figures are derived from the level 4 data downloaded from the LINCS Molecular Deep Dive files on Synapse https://www.synapse.org/#!Synapse:syn2862345/wiki/588244.  



```{r standardFigureFunctions}

#' Read an assay's csv or tsv data file 
load_data <- function(assay_name, level = 4){
  file_name <- dir(pattern = paste0("MDD_", assay_name,"_","Level",level,".csv"),recursive = TRUE, full.names = TRUE)
  if(!length(file_name) == 1) stop("There was not one ", assay_name, " data file found")
    
  if(str_detect(file_name, "csv")) {
    df <- read_csv(file_name)
  } else if (str_detect(file_name, "tsv")){
    df <- read_tsv(file_name)
  } else stop("The data file ",file_name, " must be a csv or tsv file type")
} #end of function

get_win_probs <- function(df, low = .02, hi = .98){
  windsorize_probs <- df %>%
    dplyr::select(-feature) %>%
    unlist %>%
    quantile(probs = c(low, hi), na.rm = TRUE)
}
  
prep_data <- function(df, md){
    df <- df %>%
    gather(specimenID, value, -feature) %>%
    inner_join(md, by = "specimenID") %>%
    dplyr::select(feature, specimenName, value) %>%
    spread(specimenName, value) %>%
    mutate(feature = factor(feature, levels = unique(df$feature))) %>%
      arrange(feature,)
  #Set the order of the columns
  condition_order <-  paste(rep(c("ctrl","PBS", "HGF", "OSM", "EGF","BMP2", "IFNG", "TGFB"), each = 30, times = 3), rep(c(0, 1, 4 ,8, 24, 48), each = 5, times = 3), rep(c("C1", "C2", "C3"), each = 240), rep(c("A", "B", "C", "D", "E"), times = 109),  sep = "_") %>%
    intersect(colnames(df))
  
  #Arrange columns by standard condition order
  df <- df %>%
    dplyr::select(feature, condition_order)
  return(df)
}

prep_annotations <- function(df, md){
      df_ann <- df %>%
      gather("specimenName", "value", -feature) %>%
      inner_join(md, by = "specimenName") %>%
      dplyr::select(ligand, secondLigand, experimentalTimePoint, replicate, collection) %>%
      distinct() %>%
      rename(Time = experimentalTimePoint,
             Ligand = ligand,
             Ligand2 = secondLigand,
             Replicate = replicate,
             Collection = collection) %>%
      as.data.frame()
      return(df_ann)
}
 #Create the heatmap column annotations
create_top_annotations <- function(df, md){
  df_ann <- prep_annotations(df, md)
  haTop <- HeatmapAnnotation(df = df_ann,
                             col = list(Ligand = c("ctrl" = "#7A4A2A",
                                                   "PBS" = "#8dd3c7",
                                                    "HGF" = "#80b1d3",
                                                    "OSM" = "#fdb462",
                                                    "EGF" = "#fb8072",
                                                    "BMP2" = "#b3de69",
                                                    "IFNG" = "#bebada",
                                                    "TGFB" = "#ffd92f"),
                                        Time = c("0" = "white", "1" = "azure1", "4" = "azure2", "8" = "azure3", "24" = "azure4", "48" = "black"), 
                                        Ligand2 = c("EGF" = "#9E79A3", "none" = "#BE7249"),
                                        Replicate = c("A" = '#80b1d3',"B" = '#fdb462',"C" = '#ffd92f',"D" = '#8dd3c7',"E" = '#bebada'),
                                        Collection = c("C1" = "#D79FD4", "C2" = "#3CC3B9", "C3" = "#B9B34C")))
  return(haTop)
}

prep_top_annotations <- function(df, md){
  df_ann <- prep_annotations(df, md)
  haTop <- HeatmapAnnotation(df = df_ann,
                             col = list(Ligand = c("ctrl" = "#7A4A2A",
                                                   "PBS" = "#8dd3c7",
                                                    "HGF" = "#80b1d3",
                                                    "OSM" = "#fdb462",
                                                    "EGF" = "#fb8072",
                                                    "BMP2" = "#b3de69",
                                                    "IFNG" = "#bebada",
                                                    "TGFB" = "#ffd92f"),
                                        Time = c("0" = "white", "1" = "azure1", "4" = "azure2", "8" = "azure3", "24" = "azure4", "48" = "black"), 
                                        Ligand2 = c("EGF" = "#9E79A3", "none" = "#BE7249"),
                                        Replicate = c("A" = '#80b1d3',"B" = '#fdb462',"C" = '#ffd92f',"D" = '#8dd3c7',"E" = '#bebada'),
                                        Collection = c("C1" = "#D79FD4", "C2" = "#3CC3B9", "C3" = "#B9B34C")))
  return(haTop)
}

create_row_annotations <- function(df, md){
  df_ann <- prep_annotations(df, md) %>%
    dplyr::select(Collection, Replicate, Time, Ligand2, Ligand)
  haRow <- rowAnnotation(df = df_ann, show_legend = FALSE,
                             col = list(Ligand = c( "ctrl" = "#7A4A2A",
                                                    "PBS" = "#8dd3c7",
                                                    "HGF" = "#80b1d3",
                                                    "OSM" = "#fdb462",
                                                    "EGF" = "#fb8072",
                                                    "BMP2" = "#b3de69",
                                                    "IFNG" = "#bebada",
                                                    "TGFB" = "#ffd92f"),
                                        Time = c("0" = "white", "1" = "azure1", "4" = "azure2", "8" = "azure3", "24" = "azure4", "48" = "black"), 
                                        Ligand2 = c("EGF" = "#9E79A3", "none" = "#BE7249"),
                                        Replicate = c("A" = '#80b1d3',"B" = '#fdb462',"C" = '#ffd92f',"D" = '#8dd3c7',"E" = '#bebada'),
                                        Collection = c("C1" = "#D79FD4", "C2" = "#3CC3B9", "C3" = "#B9B34C")))
  return(haRow)
}

  
plot_heatmap <- function(df, md, assay_name, EGF_normed = TRUE, var_thresh_percentile = 0, show_row_names = FALSE, mark_rows = NULL, cluster_rows = TRUE, select_timepoint = FALSE, zscore_rows = TRUE,...) {
  
  df_vars <- df %>%
  dplyr::select(-feature) %>%
  as.matrix() %>%
  apply(1, var, na.rm = TRUE)
  
var_thresh <- quantile(df_vars, probs = var_thresh_percentile, na.rm = TRUE)
df <- df[df_vars >= var_thresh,]
  
  #convert specimenID column names to specimanName and reorder
  df <- prep_data(df, md)
title_suffix <- ""
  if(EGF_normed){
    title_suffix = ", EGF timecourse normalized"
    df <- df %>%
    dplyr::select(-matches("EGF"))
  }

  if(select_timepoint){
    df <- df %>%
    dplyr::select(matches(paste0("_",select_timepoint,"_","|feature")))
  }

  if(!is.null(mark_rows)) {
    row_nums <- which(df$feature %in% mark_rows)
    ra <- rowAnnotation(foo = anno_mark(at = c(row_nums), labels = df$feature[row_nums]))
  } else {
    ra <- NULL
  }
if(zscore_rows) {
  df_as_matrix <- dplyr::select(df, -feature) %>%
    as.matrix() %>%
    t %>%
    scale() %>%
    t
} else {
  df_as_matrix <- dplyr::select(df, -feature) %>%
    as.matrix()
}
  rownames(df_as_matrix) <- df$feature

 windsorize_probs <- df_as_matrix %>%
      unlist %>% 
   quantile(probs = c(.02, .98), na.rm = TRUE)
 
   #Create the heatmap
  hm <- Heatmap(matrix = df_as_matrix,
          name = "abundance",
          column_title = paste0(assay_name, title_suffix),
          top_annotation = create_top_annotations(df, md),
          right_annotation = ra,
          show_row_names = show_row_names,
          show_column_names = FALSE,
          cluster_columns = FALSE,
          cluster_rows = cluster_rows,
          col = circlize::colorRamp2(c(windsorize_probs[1], 0, windsorize_probs[2]), c("blue", "white", "red")),
          na_col = "grey",...)
  hm
}

plot_line_graphs <- function(df, md, ligand_cols, EGF_normed = TRUE, fc_thresh = 1, assay_name){
  
  windsorize_probs <- get_win_probs(df, .01, .99)
  
  df_sum <- df %>%
    prep_data(md = md) %>%
    gather("specimenName", "value", -feature) %>%
    inner_join(md, by = "specimenName") %>%
    dplyr::select(ligand, secondLigand, experimentalTimePoint, replicate, collection, feature, value) %>%
    rename(Time = experimentalTimePoint,
           Ligand = ligand,
           Ligand2 = secondLigand,
           Replicate = replicate,
           Collection = collection) %>%
    group_by(feature, Collection, Time, Ligand) %>%
    summarise(value = median(value)) %>%
    ungroup() 
  
  if(0 %in% unique(df_sum$Time)){
    df2 <- df_sum %>%
      filter(Time == 0) %>%
      data.frame(Ligand = rep(unique(df_sum$Ligand), each = length(unique(df_sum$feature))),
                 stringsAsFactors = FALSE) %>%
      dplyr::select(feature, Ligand.1, value, Collection, Time) %>%
      rename(Ligand = Ligand.1) %>%
      bind_rows(df_sum) %>%
      filter(!Ligand == "ctrl") 
  } else {
    df2 <- df_sum
  }
  
  
  fc_set <- df2 %>%
    group_by(Time, Ligand) %>%
    filter(abs(value) > fc_thresh) %>%
    ungroup() %>%
    dplyr::select(feature) %>%
    distinct() %>%
    inner_join(df2, by = "feature") %>%
    mutate(Ligand = factor(Ligand, levels = names(ligand_cols)))
  
  title_suffix <- ""
  if(EGF_normed){
    title_suffix <- ", EGF normalized"
    filter(fc_set, !ligand == "EGF")
  }
  p <- ggplot(fc_set, aes(x=Time, y=value, colour=Ligand))+
    geom_line()+
    labs(title=paste0("Line Graphs",title_suffix),
         x = "Time (hours)",
         y="Intensity (AU)") +
    scale_x_continuous(breaks = c(0,8,24,48)) +
    scale_color_manual(values = ligand_cols) +
    theme(axis.text.x = element_text(angle=90),
          axis.text.y=element_blank(),
          axis.ticks.y=element_blank(),
          strip.text = element_text(size=10, hjust=0),
            panel.background = element_rect(fill = NA),
  panel.grid.major = element_line(colour = "gray95"))+
    facet_wrap(~feature, ncol = 10,scales = "free_y")
  p
  
  }
  
plot_umap <- function(df, md, ligand_cols = ligand_cols, assay_name,
                   ligand_2_cols = c("EGF" = "#ff0000",
                     "None" = "#00000000"), select_timepoint = FALSE, ...){
  #Generate a 2d scatterplot of the UMAP dimensions
  #Use size to represent timepoint, dot in center to represent ligand_2 == EGF

  windsorize_probs <- get_win_probs(df, .01, .99)
  
  #convert specimenID column names to specimanName and reorder
  df <- prep_data(df, md) %>%
    dplyr::select(-matches("EGF"))
  
    if(select_timepoint){
    df <- df %>%
    dplyr::select(matches(paste0("_",select_timepoint,"_","|feature")))
  }

   #Create annotation values
  df_ann <-prep_annotations(df, md)

   df_mat <- df %>%
    dplyr::select(-feature) %>%
     t() %>%
  as.matrix()
  colnames(df_mat) <- df$feature

  df_UMAP <- umap(df_mat, na.rm = TRUE, ...)$layout %>%
  data.frame() %>%
  rename(UMAP_1 = X1,
         UMAP_2 = X2) %>%
    cbind(df_ann) %>%
    mutate(Ligand = factor(Ligand, levels = names(ligand_cols)))
  
p <- ggplot(df_UMAP, aes(x = UMAP_1,
                         y = UMAP_2,
                         size = Time,
                         colour = Ligand,
                         fill = Ligand2)) +
    geom_point(alpha = .8) +
  geom_point(shape = 21, stroke = 4, alpha = .8) +
  scale_color_manual( values = ligand_cols) +
  scale_fill_manual( values = ligand_2_cols) +
  scale_size(breaks = c(1, 4,8,24,48)) +
  labs(title = paste("UMAP embedding of ", assay_name, " data")) +
  theme_bw()
p
  
}

plot_pca <- function(df, md, ligand_cols = ligand_cols, assay_name,
                   ligand_2_cols = c("EGF" = "#ff0000",
                     "None" = "#00000000")){
  #Generate a 2d scatterplot of the first two princpal components
  #Use size to represent timepoint, dot in center to represent ligand_2 == EGF

  windsorize_probs <- get_win_probs(df, .01, .99)
  
  #convert specimenID column names to specimenName and reorder
  df <- prep_data(df, md) %>%
    dplyr::select(-matches("EGF"))
  
   #Create annotation values
  df_ann <-prep_annotations(df, md)

   df_mat <- df %>%
    dplyr::select(-feature) %>%
     t() %>%
  as.matrix()
  colnames(df_mat) <- df$feature

    df_pca <- prcomp(df_mat)$x %>%
  data.frame() %>%
    cbind(df_ann)
    
  df_pca <- prcomp(df_mat)$x %>%
  data.frame() %>%
    cbind(df_ann) %>%
    mutate(Ligand = factor(Ligand, levels = names(ligand_cols)))
  
p <- ggplot(df_pca, aes(x = PC1,
                         y = PC2,
                         size = Time,
                         colour = Ligand,
                         fill = Ligand2)) +
  geom_point(alpha = .8) +
  geom_point(shape = 21, stroke = 4, alpha = .8) +
  scale_color_manual( values = ligand_cols) +
  scale_fill_manual( values = ligand_2_cols) +
  scale_size_area(max_size = 10, breaks = c(1,4,8,24,48)) +
  labs(title = paste("PCA of ", assay_name, " data")) + 
  theme_bw()
p
}


plot_correlation <- function(df, md, assay_name, EGF_normed = TRUE,
                      ligand_cols = c( "PBS" = "#8dd3c7",
                                      "HGF" = "#80b1d3",
                                      "OSM" = "#fdb462",
                                      "EGF" = "#fb8072",
                                      "BMP2" = "#b3de69",
                                      "IFNG" = "#bebada",
                                      "TGFB" = "#ffd92f"),
                   ligand_2_cols = c("EGF" = "#ff0000",
                     "None" = "#00000000")){
  #calculate correlations across the conditions and show in a heatmap

  #convert specimenID column names to specimanName, reorder delete EGF samples
  df <- prep_data(df, md) 
  
  title_suffix <- ""
  if(EGF_normed){
    title_suffix <- ", EGF Normalized"
    df <- df %>%
    dplyr::select(-matches("EGF"))
  }
  #Create annotation values
  #df_ann <-prep_annotations(df, md)
  
  haRow <- create_row_annotations(df, md)
  
  #Create the heatmap
  Heatmap(matrix = dplyr::select(df, -feature) %>%
            as.matrix() %>%
            t %>%
            scale(scale = FALSE) %>%
            t %>%
            cor(use = "complete.obs",method = "spearman"),
          name = "correlation",
          column_title = paste0("Correlation of ",assay_name, title_suffix),
          top_annotation = create_top_annotations(df, md),
          left_annotation = haRow,
          show_row_names = FALSE,
          show_column_names = FALSE,
          cluster_rows = FALSE,
          cluster_columns = FALSE,
          col = circlize::colorRamp2(c(-1, 0, 1), c("blue", "white", "red")),
          na_col = "grey")
}

##########
```

```{r globalVariables}
ligand_cols <- c("ctrl" = "#7A4A2A",
                 "PBS" = "#8dd3c7",
                  "HGF" = "#80b1d3",
                  "OSM" = "#fdb462",
                  "EGF" = "#fb8072",
                  "BMP2" = "#b3de69",
                  "IFNG" = "#bebada",
                  "TGFB" = "#ffd92f")

ligand_2_cols = c("EGF" = "#ff0000",
                     "None" = "#00000000")

md <- read_csv("metadata/MDD_sample_annotations.csv")
```

### RPPA{.tabset .tabset-fade}

#### Heatmap

```{r rppa_figures}

RPPA_l1 <- load_data(assay_name = "RPPA", level = 1) %>%
  rename(feature = antibody) %>%
  gather(key = specimenID, value = value, -feature) %>%
  group_by(feature) %>%
  mutate(value = value - median(value, na.rm = TRUE)) %>%
  ungroup() %>%
    spread(key = specimenID, value = value)

```


```{r}
var_thresh_percentile <- 0

plot_heatmap(RPPA_l1, md, assay_name = "RPPA", EGF_normed = FALSE, var_thresh_percentile = var_thresh_percentile, mark_rows = c("Caspase-7-cleaved", "c-Myc",  "Connexin-43", "DUSP4", "EGFR", "EMA", "HER2", "IRF-1", "PD-L1", "Stat3_pY705"))

```

#### Line graphs

```{r, fig.height=30, fig.width=15}

fc_thresh = 0

plot_line_graphs(df = RPPA_l1, md, ligand_cols = ligand_cols, EGF_normed = FALSE, fc_thresh = fc_thresh, assay_name = "RPPA")

```


#### UMAP

```{r}

plot_umap(RPPA_l1, md, ligand_cols = ligand_cols, assay_name = "RPPA")

```

#### PCA 

```{r}

plot_pca(RPPA_l1, md, ligand_cols = ligand_cols, assay_name = "RPPA")

```

#### Correlation

```{r}

#plot_correlation(RPPA, md, assay_name = "RPPA")

var_thresh_percentile <- .5

RPPA_l1_vars <- RPPA_l1 %>%
  dplyr::select(-feature) %>%
  as.matrix() %>%
  apply(1, var, na.rm = TRUE)
  
var_thresh <- quantile(RPPA_l1_vars, probs = var_thresh_percentile)
RPPA_l1_hi_vars <- RPPA_l1[RPPA_l1_vars > var_thresh,]

plot_correlation(RPPA_l1_hi_vars, md, EGF_normed = FALSE, assay_name = "RPPA")

```

Spearman (rank-based) correlations using all available pairwise data points.  

#### Pathways

```{r, fig.height=6}

RPPA_l1_pathways <- load_data("RPPA", "1_Pathways_Score") %>%
  rename(specimenID = X1) %>%
  gather(key = feature, value = Score, -specimenID) %>%
  spread(key = specimenID, value = Score)

plot_heatmap(RPPA_l1_pathways, md, EGF_normed = FALSE, assay_name = "RPPA",show_row_names = TRUE,
                       row_names_gp = gpar(fontsize = 10))

#pathway line graphs
  df_sum <- RPPA_l1_pathways %>%
    prep_data(md = md) %>%
    gather("specimenName", "value", -feature) %>%
    inner_join(md, by = "specimenName") %>%
    dplyr::select(ligand, secondLigand, experimentalTimePoint, replicate, collection, feature, value) %>%
    rename(Time = experimentalTimePoint,
           Ligand = ligand,
           Ligand2 = secondLigand,
           Replicate = replicate,
           Collection = collection) %>%
    group_by(feature, Collection, Time, Ligand) %>%
    summarise(value = median(value)) %>%
    ungroup() 
  
   if(0 %in% unique(df_sum$Time)){
    df2 <- df_sum %>%
      filter(Time == 0) %>%
      data.frame(Ligand = rep(unique(df_sum$Ligand), each = length(unique(df_sum$feature))),
                 stringsAsFactors = FALSE) %>%
      dplyr::select(feature, Ligand.1, value, Collection, Time) %>%
      rename(Ligand = Ligand.1) %>%
      bind_rows(df_sum) %>%
      filter(!Ligand == "ctrl") 
  } else {
    df2 <- df_sum
  }
  
  
  fc_set <- df2 %>%
    dplyr::select(feature) %>%
    distinct() %>%
    inner_join(df2, by = "feature") %>%
    mutate(Ligand = factor(Ligand, levels = names(ligand_cols)))
  

  
p <- ggplot(fc_set, aes(x=Time, y=value, colour=Ligand))+
  geom_line()+
  labs(title=paste0("RPPA pathways"),
       x = "Time (hours)",
       y="Intensity (AU)") +
  scale_x_continuous(breaks = c(0,8,24,48)) +
  scale_color_manual(values = ligand_cols) +
  theme(axis.text.x = element_text(angle=90),
        # axis.text.y=element_blank(),
        # axis.ticks.y=element_blank(),
        strip.text = element_text(size=6, hjust=0),
        panel.background = element_rect(fill = NA),
        panel.grid.major = element_line(colour = "gray95"))+
  facet_wrap(~feature)
p
  
```


Pathway method summary  

Pathways are scored by assigning some of the proteins to pathways with weights of +/-1, summing the products of the weights and the RPPA norm linear values and then dividing by the number of proteins assigned to each pathway. The scores are displayed as Z scores within each row.  
  
<bksp>  
<bksp>  

<bksp>  

```{r pathwayComponents, fig.width=10, fig.height=6}
library(plotly)

RPPA_pathway_components <- read_tsv("../RPPAPathway/Pathway_Scores_MDD_RPPA_Level1.txt") %>%
  spread(key = Pathway, value = Weight) %>%
  dplyr::select(-Count)

RPPA_pathway_components_matrix <- RPPA_pathway_components %>%
  dplyr::select(-Predictor) %>%
  as.matrix()
rownames(RPPA_pathway_components_matrix) <- RPPA_pathway_components$Predictor

f1 <- list(
  family = "Arial, sans-serif",
  size = 12,
  color = "black"
)
f2 <- list(
  family = "Old Standard TT, serif",
  size = 12,
  color = "black"
)
a <- list(
  titlefont = f1,
  showticklabels = TRUE,
  tickangle = 0,
  tickfont = f1,
  exponentformat = "E",
  autorange = "reversed"
)

plot_ly(x=rownames(RPPA_pathway_components_matrix), y=colnames(RPPA_pathway_components_matrix),
        z = t(RPPA_pathway_components_matrix),
        type = "heatmap",
        colorscale = list(c(-1,0,1),c("blue", "white", "red"))) %>%
  layout(title = "Pathway components and weights", yaxis = a, xaxis = list(tickfont = list(size = 10)), showlegend = FALSE)


```


###  

### IF{.tabset .tabset-fade}

#### Heatmap  

```{r IF_figures, eval = TRUE}

boundedLog2 <- function(x) log2(x + .0001)

immunofluoresence <- load_data(assay_name = "IF", level = 3) %>%
  mutate_if(is.numeric, .funs = boundedLog2) %>%
  filter(!feature == "EdUPositiveProportion")
  
plot_heatmap(immunofluoresence, md, assay_name = "IF", EGF_normed = FALSE, mark_rows = immunofluoresence$feature)

```

The row features in the heatmap are `r knitr::combine_words(immunofluoresence$feature)`

#### UMAP

```{r}

plot_umap(immunofluoresence, md, ligand_cols = ligand_cols, assay_name = "IF")

```

#### PCA 

```{r}

plot_pca(immunofluoresence, md, ligand_cols = ligand_cols, assay_name = "IF")

```

#### Correlation

```{r}

plot_correlation(immunofluoresence, md, assay_name = "IF")

```

Spearman (rank-based) correlations using all available pairwise data points.  


### cycIF{.tabset .tabset-fade}

#### Heatmap 

```{r cycIF_figures, eval = TRUE}

var_thresh_percentile <- 0

cycIF <- load_data(assay_name = "cycIF", level = 3) %>%
  filter(str_detect(feature, "_int_"),
         !str_detect(feature, "none|dna")) %>%
  mutate(feature = str_remove(feature, "_.*int_mean"),
         feature = str_replace(feature,"human",""),
         feature = str_replace(feature,"alpha","α"),
         feature = str_replace(feature,"beta","β")) %>%
  mutate_if(is.numeric, log2)

cycIF <- prep_data(cycIF, md = md) %>%
  gather(specimenName, value, -feature) %>%
  mutate(Replicate = str_remove(specimenName, ".*_")) %>%
  group_by(Replicate, feature) %>%
  mutate(MedCenter = value-median(value, na.rm = TRUE),
         ZScore = MedCenter/sd(value, na.rm = TRUE)) %>%
  ungroup() %>%
    left_join(md) %>%
  dplyr::select(feature, specimenID, ZScore) %>%
  spread(key = specimenID, value = ZScore)

cycIF_vars <- cycIF %>%
  dplyr::select(-feature) %>%
  as.matrix() %>%
  apply(1, var)
  
var_thresh <- quantile(cycIF_vars, probs = var_thresh_percentile)
cycIF_hi_vars <- cycIF[cycIF_vars >= var_thresh,]

cycIF_hm <- plot_heatmap(cycIF_hi_vars, md, assay_name = "cycIF", EGF_normed = FALSE,show_row_names = TRUE, row_names_gp = gpar(fontsize = 7))

print(cycIF_hm)
```

All measurements are the mean intensities of the antibody in the labeled compartment. Values are Z scores within each replicate set.


#### Line graphs

```{r, fig.height=15, fig.width=15}

plot_line_graphs(cycIF, md, EGF_normed = FALSE, ligand_cols = ligand_cols, assay_name = "cycIF")

```


#### UMAP

```{r}

plot_umap(cycIF, md, ligand_cols = ligand_cols, assay_name = "cycIF")

```

#### PCA 

```{r}

plot_pca(cycIF, md, ligand_cols = ligand_cols, assay_name = "cycIF")

```

#### Correlation

```{r}

plot_correlation(cycIF, md, EGF_normed = FALSE, assay_name = "cycIF")

```

Spearman (rank-based) correlations using all available pairwise data points.  


### RNAseq{.tabset .tabset-fade}

#### Heatmap  

```{r rnaSeq_figures, eval = TRUE}
RNAseq <- load_data(assay_name = "RNAseq",level = 2) %>%
  rename(feature = ensembl_gene_id)

RNAseq_vars <- RNAseq %>%
  dplyr::select(-feature) %>%
  as.matrix() %>%
  apply(1, var)
  
var_thresh_percentile <- .95
var_thresh <- quantile(RNAseq_vars, probs = var_thresh_percentile)
RNAseq_hi_vars <- RNAseq[RNAseq_vars > var_thresh,]
  
plot_heatmap(RNAseq_hi_vars, md, assay_name = "RNAseq",EGF_normed = FALSE)

```


Genes included in the heatmap have a variance above the `r var_thresh_percentile` percentile . 

#### UMAP

```{r}

plot_umap(RNAseq, md, ligand_cols = ligand_cols, assay_name = "RNAseq")

```

#### PCA 

```{r}

plot_pca(RNAseq, md, ligand_cols = ligand_cols, assay_name = "RNAseq")

```

#### Correlation

```{r}

plot_correlation(RNAseq_hi_vars, md, assay_name = "RNAseq",EGF_normed = FALSE)

```

Spearman (rank-based) correlations using all available pairwise data points.  

```{r writeOutChea3InputFiles, eval = FALSE}

#Get annotations to convert from ensemble to HGNC
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL",
               dataset = "hsapiens_gene_ensembl",
               host = "uswest.ensembl.org")

annoTable <- getBM(attributes = c("ensembl_gene_id",
                            "hgnc_symbol"),
             mart = mart)

#Get ChEA3 results for based on RNAseq level 2 Z scores for each gene

RNAseq_zscores <- dplyr::select(RNAseq, -feature) %>%
  as.matrix() %>%
  t %>%
  scale() %>%
  t %>%
  as.tibble() %>%
  mutate(feature = RNAseq$feature) %>%
  left_join(annoTable, by = c("feature" = "ensembl_gene_id")) %>%
  dplyr::select(-feature) %>%
  rename(feature = hgnc_symbol) %>%
  gather(specimenID, value, -feature) %>%
  inner_join(md, by = "specimenID")

#write out high and low scored genes for each condition
for(specName in unique(RNAseq_zscores$specimenName)){
  foo <- RNAseq_zscores %>%
  filter(specimenName == specName,
         value > 2.5) %>%
    dplyr::select(feature) %>%
    distinct %>%
    write_csv(path = paste0("RNAseq/Data/",specName,"_extreme_genes.txt"), col_names = FALSE)
}

```

```{r getChEA3TFs}
#read in TF scores
getTFs <- function(x){
  library(httr)
  library(jsonlite)
  genes <- read_tsv(paste0("./RNAseq/Data/",x)) %>%
    distinct %>%
    unlist 
  genes <- genes[!is.na(genes)]
  
  url = "https://amp.pharm.mssm.edu/chea3/api/enrich/"
  encode = "json"
  payload = list(query_name = x, gene_set = genes)
  
  #POST to ChEA3 server
  response = POST(url = url, body = payload, encode = encode)
  json = content(response, "text")
  
  #results as list of R dataframes
  results <- try(jsonlite::fromJSON(json))
  if(!is.null(results)) {
    results <- results[["Integrated--meanRank"]]
  } 
  return(results)
}

if(!file.exists("RNAseq/Data/ChEA3_TFs.csv")){
  #gene_fns <- dir("RNAseq/Data/", pattern = "_[248]*h.txt")
  gene_fns <- dir("RNAseq/Data/", pattern = "extreme_genes")
  TFs <- map(gene_fns[1], getTFs) %>%
    bind_rows()  %>%
  rename(Query_Name = 'Query Name') %>%
  mutate(specimenName = str_remove(Query_Name, "_extreme_genes.txt"),
         Score = as.numeric(Score))%>%
  select(specimenName, TF, Score) %>%
  spread(key = TF, value = Score)
  write_csv(foo, "RNAseq/Data/ChEA3_TFs.csv")
} else{
  TFs <- read_csv("RNAseq/Data/ChEA3_TFs.csv")
}

```

```{r plotTFsvsPathways}

TF_pathways <- RPPA_l1_pathways %>%
    gather(specimenID, value, -feature) %>%
  inner_join(md, by = "specimenID") %>%
   select(specimenName, feature, value) %>%
  spread(key = feature, value = value) %>%
  inner_join(TFs) %>%
  gather(key = feature, value = "value", -specimenName) %>%
   inner_join(md, by = "specimenName") %>%
     select(specimenID, feature, value) %>%
  spread(key = specimenID, value = value)
  
plot_heatmap(TF_pathways, md, assay_name = "TFs and pathways",EGF_normed = FALSE,show_row_names = TRUE)

```



### GCP{.tabset .tabset-fade}

#### Heatmap  

```{r GCP_figures}

var_thresh_percentile <- 0

GCP_level2 <- load_data(assay_name = "GCP", level = 2) %>%
  rename(feature = histone) %>%
  prep_data(md = md) %>%
  gather(specimenName, value, -feature) %>%
  mutate(Replicate = str_remove(specimenName, ".*_")) %>%
  group_by(Replicate, feature) %>%
  mutate(MedCenter = value-median(value, na.rm = TRUE),
         ZScore = MedCenter/sd(value, na.rm = TRUE)) %>%
  ungroup() %>%
    left_join(md) %>%
  dplyr::select(feature, specimenID, ZScore) %>%
  spread(key = specimenID, value = ZScore) %>%
  dplyr::select(-sid255, -sid256)

plot_heatmap(GCP_level2, md, assay_name = "GCP", EGF_normed = FALSE, var_thresh_percentile = var_thresh_percentile, show_row_names = TRUE, cluster_rows = FALSE)

```


#### Line graphs

```{r, fig.height=15, fig.width=15}

plot_line_graphs(GCP_level2, md, ligand_cols = ligand_cols, assay_name = "GCP", EGF_normed = FALSE, fc_thresh = 0)

```

#### UMAP

```{r}
df <- GCP_level2 %>%
  drop_na()
plot_umap(df, md, ligand_cols = ligand_cols, assay_name = "GCP")

```

#### PCA 

```{r}
df <- GCP_level2 %>%
  drop_na()
plot_pca(df, md, ligand_cols = ligand_cols, assay_name = "GCP")

```

#### Correlation

```{r}

var_thresh_percentile <- .5

GCP_vars <- GCP_level2 %>%
  dplyr::select(-feature) %>%
  as.matrix() %>%
  apply(1, var, na.rm = TRUE)
  
var_thresh <- quantile(GCP_vars, probs = var_thresh_percentile)
GCP_hi_vars <- GCP_level2[GCP_vars > var_thresh,]

plot_correlation(GCP_hi_vars, md, assay_name = "GCP",EGF_normed = FALSE)

GCP_level2_vars <- GCP_level2 %>%
  dplyr::select(-feature) %>%
  as.matrix() %>%
  apply(1, var, na.rm = TRUE)
  
var_thresh <- quantile(GCP_level2_vars, probs = var_thresh_percentile)
GCP_level2_hi_vars <- GCP_level2[GCP_level2_vars > var_thresh,]

countNAs <- GCP_level2_hi_vars %>%
  mutate_all(is.na) %>%
  summarise_all(sum) %>%
  unlist

GCP_level2_hi_vars <- GCP_level2_hi_vars[,!countNAs==nrow(GCP_level2_hi_vars)]

plot_correlation(GCP_level2_hi_vars, md, EGF_normed = FALSE, assay_name = "GCP")

```

Spearman (rank-based) correlations using all available pairwise data points.  

The heatmap shows the spearman correlation values of the top `r var_thresh_percentile*100`th percentile varying features.  

### ATACseq{.tabset .tabset-fade}

#### Heatmap  

```{r ATACseq_figures, eval = TRUE}

ATACseq <- load_data(assay_name = "ATACseq", level = 2) %>%
  rename(feature = peak)

ATACseq_vars <- ATACseq %>%
  dplyr::select(-feature) %>%
  as.matrix() %>%
  apply(1, var)
  
var_thresh_percentile <- .98
var_thresh <- quantile(ATACseq_vars, probs = var_thresh_percentile)
ATACseq_hi_vars <- ATACseq[ATACseq_vars > var_thresh,]
  
plot_heatmap(ATACseq_hi_vars, md, EGF_normed = FALSE, assay_name = "ATACseq high variance peaks")

```

Peaks included in the heatmap have a variance above the `r var_thresh_percentile` percentile . 


```{r ATACseq_chr1}

ATACseq_chr1 <- ATACseq %>%
  filter(str_detect(feature,"chr1_"))

plot_heatmap(ATACseq_chr1, md, EGF_normed = FALSE, assay_name = "ATACseq chromosome 1 peaks")

```



#### UMAP

```{r}

plot_umap(ATACseq, md, ligand_cols = ligand_cols, assay_name = "ATACseq")

```

The UMAP embedding is based on `r nrow(ATACseq)` peaks for each sample.  

#### PCA 

```{r}

plot_pca(ATACseq, md, ligand_cols = ligand_cols, assay_name = "ATACseq")

```

PCA is based on `r nrow(ATACseq)` peaks for each sample.  

#### Correlation

```{r}

plot_correlation(ATACseq_hi_vars, md, assay_name = "ATACseq",EGF_normed = FALSE)

```

Spearman (rank-based) correlations using all available pairwise data points.  


### L1000{.tabset .tabset-fade}

#### Heatmap  
 
```{r l1000_figures}

var_thresh_percentile <- 0

L1000 <- load_data("L1000", level = 2) %>%
  rename(feature = probeset) %>%
  gather(key = specimenID, value = value, -feature) %>%
  left_join(md) %>%
  dplyr::select(feature, specimenID, value, collection) %>%
  group_by(feature, collection) %>%
  mutate(value = value - median(value, na.rm = TRUE)) %>%
  ungroup() %>%
  dplyr::select(-collection) %>%
  spread(key = specimenID, value = value)

```


```{r}

plot_heatmap(L1000, md, "L1000",EGF_normed = FALSE, var_thresh_percentile = var_thresh_percentile)

L1000_c1 <- L1000 %>%
  gather(key = specimenID, value = value, -feature) %>%
  left_join(md) %>%
  dplyr::select(feature, specimenID, value, collection) %>%
  filter(collection == "C1") %>%
  group_by(feature) %>%
  mutate(value = value - median(value, na.rm = TRUE)) %>%
  ungroup() %>%
  dplyr::select(-collection) %>%
  spread(key = specimenID, value = value)

plot_heatmap(L1000_c1, md, "L1000 collection 1",EGF_normed = FALSE, var_thresh_percentile = var_thresh_percentile)

L1000_c2 <- L1000 %>%
  gather(key = specimenID, value = value, -feature) %>%
  left_join(md) %>%
  dplyr::select(feature, specimenID, value, collection) %>%
  filter(collection == "C2") %>%
  group_by(feature) %>%
  mutate(value = value - median(value, na.rm = TRUE)) %>%
  ungroup() %>%
  dplyr::select(-collection) %>%
  spread(key = specimenID, value = value)

plot_heatmap(L1000_c2, md, "L1000 collection 2",EGF_normed = FALSE, var_thresh_percentile = var_thresh_percentile)

L1000_c3 <- L1000 %>%
  gather(key = specimenID, value = value, -feature) %>%
  left_join(md) %>%
  dplyr::select(feature, specimenID, value, collection) %>%
  filter(collection == "C3") %>%
  group_by(feature) %>%
  mutate(value = value - median(value, na.rm = TRUE)) %>%
  ungroup() %>%
  dplyr::select(-collection) %>%
  spread(key = specimenID, value = value)

plot_heatmap(L1000_c1, md, "L1000 collection 3",EGF_normed = FALSE, var_thresh_percentile = var_thresh_percentile)
```


#### UMAP

```{r, fig.height=8, fig.width=9}

L1000_C1_C2 <- L1000 %>%
  gather(key = specimenID, value = value, -feature) %>%
  left_join(md) %>%
  dplyr::select(feature, specimenID, value, collection) %>%
  filter(collection %in% c("C1", "C2")) %>%
  group_by(feature, collection) %>%
  mutate(value = value - median(value, na.rm = TRUE)) %>%
  ungroup() %>%
  dplyr::select(-collection) %>%
  spread(key = specimenID, value = value)

plot_umap(L1000_C1_C2, md, ligand_cols = ligand_cols, assay_name = "L1000, C1 and C2", n_epochs = 200, min_dist = .01, n_neighbors = 15, random_state = 42)

df <- L1000
assay_name <- "L1000 T48"
  #Use size to represent timepoint, dot in center to represent ligand_2 == EGF

  windsorize_probs <- get_win_probs(df, .01, .99)
  
  #convert specimenID column names to specimanName and reorder
  df <- prep_data(df, md) %>%
    dplyr::select(matches("_48_|feature"))
 
   #Create annotation values
  df_ann <-prep_annotations(df, md)

   df_mat <- df %>%
    dplyr::select(-feature) %>%
     t() %>%
  as.matrix()
  colnames(df_mat) <- df$feature

  df_UMAP <- umap(df_mat, na.rm = TRUE, n_epochs = 200, min_dist = .01, n_neighbors = 15, random_state = 42)$layout %>%
  data.frame() %>%
  rename(UMAP_1 = X1,
         UMAP_2 = X2) %>%
    cbind(df_ann) %>%
    mutate(Ligand = factor(Ligand, levels = names(ligand_cols)))
  
p <- ggplot(df_UMAP, aes(x = UMAP_1,
                         y = UMAP_2,
                         colour = Ligand,
                         fill = Ligand2,
                         shape = Collection)) +
  geom_point(size = 12, alpha = .8) +
  geom_point(shape = 21, size = 2, alpha = .8) +
  scale_color_manual( values = ligand_cols) +
  scale_fill_manual( values = ligand_2_cols) +
  labs(title = paste("UMAP embedding of ", assay_name, " data")) +
  theme_bw()
p


```


```{r, fig.height=3, fig.width=3, eval = FALSE}

df <- L1000
  #Use size to represent timepoint, dot in center to represent ligand_2 == EGF

  windsorize_probs <- get_win_probs(df, .01, .99)
  
  #convert specimenID column names to specimanName and reorder
  df <- prep_data(df, md) %>%
    dplyr::select(matches("_48_|feature"))
 
   #Create annotation values
  df_ann <-prep_annotations(df, md)

   df_mat <- df %>%
    dplyr::select(-feature) %>%
     t() %>%
  as.matrix()
  colnames(df_mat) <- df$feature


  min_dists <- c(.01)
  n_epochs_s <- c(200)
  n_neighbors_s <- c(15)
  
  plot_df_UMAP <- function(min_dist, n_epochs, n_neighbors){
    df_UMAP <- umap(df_mat, na.rm = TRUE, n_epochs = n_epochs, min_dist = min_dist, n_neighbors = n_neighbors, random_state = 42)$layout %>%
  data.frame() %>%
  rename(UMAP_1 = X1,
         UMAP_2 = X2) %>%
    cbind(df_ann) %>%
    mutate(Ligand = factor(Ligand, levels = names(ligand_cols)))
  
p <- ggplot(df_UMAP, aes(x = UMAP_1,
                         y = UMAP_2,
                         colour = Ligand,
                         fill = Ligand2,
                         shape = Collection)) +
  geom_point(size = 5, alpha = .8) +
  geom_point(shape = 21, size = 2, alpha = .8) +
  scale_color_manual( values = ligand_cols) +
  scale_fill_manual( values = ligand_2_cols) +
  coord_cartesian(x = c(-3.2, 3.2), y = c(-3.2, 3.2)) +
  labs(title = paste("UMAP embedding of L1000 T48 data\nmin_dist =",min_dist,"epochs =",n_epochs, "knn=",n_neighbors)) +
  guides(colour = FALSE,
         fill = FALSE,
         size = FALSE,
         shape = FALSE) +
  theme_bw() +
  theme(title = element_text(size = 3))
print(p)
  }
  
  for(n_epochs in n_epochs_s){
    for(min_dist in min_dists){
      for(n_neighbors in n_neighbors_s){
      plot_df_UMAP(min_dist, n_epochs, n_neighbors)
    }
    }
  }
  
```


#### PCA 

```{r}

plot_pca(L1000, md, ligand_cols = ligand_cols, assay_name = "L1000")

```

#### Correlation

```{r}

plot_correlation(L1000, md, EGF_normed = FALSE, assay_name = "L1000")

```

Spearman (rank-based) correlations using all available pairwise data points.  


 
##